// @livekit/components-react/dist/contexts-DjicKJUz.mjs
import { RoomEvent as b, ParticipantEvent as y, setLogLevel as Fn, LogLevel as mt, setLogExtension as jn, LocalParticipant as ot, Track as R, Room as gt, TrackEvent as Ft } from "livekit-client";
import * as D from "react";
const Ke = Math.min, ue = Math.max, ke = Math.round, ee = (e) => ({
  x: e,
  y: e
}), Wn = {
  left: "right",
  right: "left",
  bottom: "top",
  top: "bottom"
}, Bn = {
  start: "end",
  end: "start"
};
function bt(e, t, n) {
  return ue(e, Ke(t, n));
}
function Ne(e, t) {
  return typeof e == "function" ? e(t) : e;
}
function oe(e) {
  return e.split("-")[0];
}
function $e(e) {
  return e.split("-")[1];
}
function jt(e) {
  return e === "x" ? "y" : "x";
}
function Wt(e) {
  return e === "y" ? "height" : "width";
}
function Ue(e) {
  return ["top", "bottom"].includes(oe(e)) ? "y" : "x";
}
function Bt(e) {
  return jt(Ue(e));
}
function Vn(e, t, n) {
  n === void 0 && (n = !1);
  const r = $e(e), i = Bt(e), o = Wt(i);
  let s = i === "x" ? r === (n ? "end" : "start") ? "right" : "left" : r === "start" ? "bottom" : "top";
  return t.reference[o] > t.floating[o] && (s = Oe(s)), [s, Oe(s)];
}
function Hn(e) {
  const t = Oe(e);
  return [Qe(e), t, Qe(t)];
}
function Qe(e) {
  return e.replace(/start|end/g, (t) => Bn[t]);
}
function zn(e, t, n) {
  const r = ["left", "right"], i = ["right", "left"], o = ["top", "bottom"], s = ["bottom", "top"];
  switch (e) {
    case "top":
    case "bottom":
      return n ? t ? i : r : t ? r : i;
    case "left":
    case "right":
      return t ? o : s;
    default:
      return [];
  }
}
function Yn(e, t, n, r) {
  const i = $e(e);
  let o = zn(oe(e), n === "start", r);
  return i && (o = o.map((s) => s + "-" + i), t && (o = o.concat(o.map(Qe)))), o;
}
function Oe(e) {
  return e.replace(/left|right|bottom|top/g, (t) => Wn[t]);
}
function Gn(e) {
  return {
    top: 0,
    right: 0,
    bottom: 0,
    left: 0,
    ...e
  };
}
function Kn(e) {
  return typeof e != "number" ? Gn(e) : {
    top: e,
    right: e,
    bottom: e,
    left: e
  };
}
function Le(e) {
  const {
    x: t,
    y: n,
    width: r,
    height: i
  } = e;
  return {
    width: r,
    height: i,
    top: n,
    left: t,
    right: t + r,
    bottom: n + i,
    x: t,
    y: n
  };
}
function yt(e, t, n) {
  let {
    reference: r,
    floating: i
  } = e;
  const o = Ue(t), s = Bt(t), a = Wt(s), c = oe(t), u = o === "y", l = r.x + r.width / 2 - i.width / 2, f = r.y + r.height / 2 - i.height / 2, h = r[a] / 2 - i[a] / 2;
  let d;
  switch (c) {
    case "top":
      d = {
        x: l,
        y: r.y - i.height
      };
      break;
    case "bottom":
      d = {
        x: l,
        y: r.y + r.height
      };
      break;
    case "right":
      d = {
        x: r.x + r.width,
        y: f
      };
      break;
    case "left":
      d = {
        x: r.x - i.width,
        y: f
      };
      break;
    default:
      d = {
        x: r.x,
        y: r.y
      };
  }
  switch ($e(t)) {
    case "start":
      d[s] -= h * (n && u ? -1 : 1);
      break;
    case "end":
      d[s] += h * (n && u ? -1 : 1);
      break;
  }
  return d;
}
const Qn = async (e, t, n) => {
  const {
    placement: r = "bottom",
    strategy: i = "absolute",
    middleware: o = [],
    platform: s
  } = n, a = o.filter(Boolean), c = await (s.isRTL == null ? void 0 : s.isRTL(t));
  let u = await s.getElementRects({
    reference: e,
    floating: t,
    strategy: i
  }), {
    x: l,
    y: f
  } = yt(u, r, c), h = r, d = {}, m = 0;
  for (let p = 0; p < a.length; p++) {
    const {
      name: g,
      fn: v
    } = a[p], {
      x: S,
      y: k,
      data: _,
      reset: E
    } = await v({
      x: l,
      y: f,
      initialPlacement: r,
      placement: h,
      strategy: i,
      middlewareData: d,
      rects: u,
      platform: s,
      elements: {
        reference: e,
        floating: t
      }
    });
    l = S ?? l, f = k ?? f, d = {
      ...d,
      [g]: {
        ...d[g],
        ..._
      }
    }, E && m <= 50 && (m++, typeof E == "object" && (E.placement && (h = E.placement), E.rects && (u = E.rects === !0 ? await s.getElementRects({
      reference: e,
      floating: t,
      strategy: i
    }) : E.rects), {
      x: l,
      y: f
    } = yt(u, h, c)), p = -1);
  }
  return {
    x: l,
    y: f,
    placement: h,
    strategy: i,
    middlewareData: d
  };
};
async function Vt(e, t) {
  var n;
  t === void 0 && (t = {});
  const {
    x: r,
    y: i,
    platform: o,
    rects: s,
    elements: a,
    strategy: c
  } = e, {
    boundary: u = "clippingAncestors",
    rootBoundary: l = "viewport",
    elementContext: f = "floating",
    altBoundary: h = !1,
    padding: d = 0
  } = Ne(t, e), m = Kn(d), g = a[h ? f === "floating" ? "reference" : "floating" : f], v = Le(await o.getClippingRect({
    element: (n = await (o.isElement == null ? void 0 : o.isElement(g))) == null || n ? g : g.contextElement || await (o.getDocumentElement == null ? void 0 : o.getDocumentElement(a.floating)),
    boundary: u,
    rootBoundary: l,
    strategy: c
  })), S = f === "floating" ? {
    ...s.floating,
    x: r,
    y: i
  } : s.reference, k = await (o.getOffsetParent == null ? void 0 : o.getOffsetParent(a.floating)), _ = await (o.isElement == null ? void 0 : o.isElement(k)) ? await (o.getScale == null ? void 0 : o.getScale(k)) || {
    x: 1,
    y: 1
  } : {
    x: 1,
    y: 1
  }, E = Le(o.convertOffsetParentRelativeRectToViewportRelativeRect ? await o.convertOffsetParentRelativeRectToViewportRelativeRect({
    elements: a,
    rect: S,
    offsetParent: k,
    strategy: c
  }) : S);
  return {
    top: (v.top - E.top + m.top) / _.y,
    bottom: (E.bottom - v.bottom + m.bottom) / _.y,
    left: (v.left - E.left + m.left) / _.x,
    right: (E.right - v.right + m.right) / _.x
  };
}
const Jn = function(e) {
  return e === void 0 && (e = {}), {
    name: "flip",
    options: e,
    async fn(t) {
      var n, r;
      const {
        placement: i,
        middlewareData: o,
        rects: s,
        initialPlacement: a,
        platform: c,
        elements: u
      } = t, {
        mainAxis: l = !0,
        crossAxis: f = !0,
        fallbackPlacements: h,
        fallbackStrategy: d = "bestFit",
        fallbackAxisSideDirection: m = "none",
        flipAlignment: p = !0,
        ...g
      } = Ne(e, t);
      if ((n = o.arrow) != null && n.alignmentOffset)
        return {};
      const v = oe(i), S = oe(a) === a, k = await (c.isRTL == null ? void 0 : c.isRTL(u.floating)), _ = h || (S || !p ? [Oe(a)] : Hn(a));
      !h && m !== "none" && _.push(...Yn(a, p, m, k));
      const E = [a, ..._], re = await Vt(t, g), C = [];
      let T = ((r = o.flip) == null ? void 0 : r.overflows) || [];
      if (l && C.push(re[v]), f) {
        const x = Vn(i, s, k);
        C.push(re[x[0]], re[x[1]]);
      }
      if (T = [...T, {
        placement: i,
        overflows: C
      }], !C.every((x) => x <= 0)) {
        var I, N;
        const x = (((I = o.flip) == null ? void 0 : I.index) || 0) + 1, ie = E[x];
        if (ie)
          return {
            data: {
              index: x,
              overflows: T
            },
            reset: {
              placement: ie
            }
          };
        let q = (N = T.filter((X) => X.overflows[0] <= 0).sort((X, ce) => X.overflows[1] - ce.overflows[1])[0]) == null ? void 0 : N.placement;
        if (!q)
          switch (d) {
            case "bestFit": {
              var w;
              const X = (w = T.map((ce) => [ce.placement, ce.overflows.filter((ge) => ge > 0).reduce((ge, Un) => ge + Un, 0)]).sort((ce, ge) => ce[1] - ge[1])[0]) == null ? void 0 : w[0];
              X && (q = X);
              break;
            }
            case "initialPlacement":
              q = a;
              break;
          }
        if (i !== q)
          return {
            reset: {
              placement: q
            }
          };
      }
      return {};
    }
  };
};
async function qn(e, t) {
  const {
    placement: n,
    platform: r,
    elements: i
  } = e, o = await (r.isRTL == null ? void 0 : r.isRTL(i.floating)), s = oe(n), a = $e(n), c = Ue(n) === "y", u = ["left", "top"].includes(s) ? -1 : 1, l = o && c ? -1 : 1, f = Ne(t, e);
  let {
    mainAxis: h,
    crossAxis: d,
    alignmentAxis: m
  } = typeof f == "number" ? {
    mainAxis: f,
    crossAxis: 0,
    alignmentAxis: null
  } : {
    mainAxis: 0,
    crossAxis: 0,
    alignmentAxis: null,
    ...f
  };
  return a && typeof m == "number" && (d = a === "end" ? m * -1 : m), c ? {
    x: d * l,
    y: h * u
  } : {
    x: h * u,
    y: d * l
  };
}
const Xn = function(e) {
  return e === void 0 && (e = 0), {
    name: "offset",
    options: e,
    async fn(t) {
      var n, r;
      const {
        x: i,
        y: o,
        placement: s,
        middlewareData: a
      } = t, c = await qn(t, e);
      return s === ((n = a.offset) == null ? void 0 : n.placement) && (r = a.arrow) != null && r.alignmentOffset ? {} : {
        x: i + c.x,
        y: o + c.y,
        data: {
          ...c,
          placement: s
        }
      };
    }
  };
}, Zn = function(e) {
  return e === void 0 && (e = {}), {
    name: "shift",
    options: e,
    async fn(t) {
      const {
        x: n,
        y: r,
        placement: i
      } = t, {
        mainAxis: o = !0,
        crossAxis: s = !1,
        limiter: a = {
          fn: (g) => {
            let {
              x: v,
              y: S
            } = g;
            return {
              x: v,
              y: S
            };
          }
        },
        ...c
      } = Ne(e, t), u = {
        x: n,
        y: r
      }, l = await Vt(t, c), f = Ue(oe(i)), h = jt(f);
      let d = u[h], m = u[f];
      if (o) {
        const g = h === "y" ? "top" : "left", v = h === "y" ? "bottom" : "right", S = d + l[g], k = d - l[v];
        d = bt(S, d, k);
      }
      if (s) {
        const g = f === "y" ? "top" : "left", v = f === "y" ? "bottom" : "right", S = m + l[g], k = m - l[v];
        m = bt(S, m, k);
      }
      const p = a.fn({
        ...t,
        [h]: d,
        [f]: m
      });
      return {
        ...p,
        data: {
          x: p.x - n,
          y: p.y - r
        }
      };
    }
  };
};
function Fe() {
  return typeof window < "u";
}
function ve(e) {
  return Ht(e) ? (e.nodeName || "").toLowerCase() : "#document";
}
function F(e) {
  var t;
  return (e == null || (t = e.ownerDocument) == null ? void 0 : t.defaultView) || window;
}
function Q(e) {
  var t;
  return (t = (Ht(e) ? e.ownerDocument : e.document) || window.document) == null ? void 0 : t.documentElement;
}
function Ht(e) {
  return Fe() ? e instanceof Node || e instanceof F(e).Node : !1;
}
function H(e) {
  return Fe() ? e instanceof Element || e instanceof F(e).Element : !1;
}
function G(e) {
  return Fe() ? e instanceof HTMLElement || e instanceof F(e).HTMLElement : !1;
}
function wt(e) {
  return !Fe() || typeof ShadowRoot > "u" ? !1 : e instanceof ShadowRoot || e instanceof F(e).ShadowRoot;
}
function Ee(e) {
  const {
    overflow: t,
    overflowX: n,
    overflowY: r,
    display: i
  } = z(e);
  return /auto|scroll|overlay|hidden|clip/.test(t + r + n) && !["inline", "contents"].includes(i);
}
function er(e) {
  return ["table", "td", "th"].includes(ve(e));
}
function je(e) {
  return [":popover-open", ":modal"].some((t) => {
    try {
      return e.matches(t);
    } catch {
      return !1;
    }
  });
}
function st(e) {
  const t = at(), n = H(e) ? z(e) : e;
  return n.transform !== "none" || n.perspective !== "none" || (n.containerType ? n.containerType !== "normal" : !1) || !t && (n.backdropFilter ? n.backdropFilter !== "none" : !1) || !t && (n.filter ? n.filter !== "none" : !1) || ["transform", "perspective", "filter"].some((r) => (n.willChange || "").includes(r)) || ["paint", "layout", "strict", "content"].some((r) => (n.contain || "").includes(r));
}
function tr(e) {
  let t = te(e);
  for (; G(t) && !de(t); ) {
    if (st(t))
      return t;
    if (je(t))
      return null;
    t = te(t);
  }
  return null;
}
function at() {
  return typeof CSS > "u" || !CSS.supports ? !1 : CSS.supports("-webkit-backdrop-filter", "none");
}
function de(e) {
  return ["html", "body", "#document"].includes(ve(e));
}
function z(e) {
  return F(e).getComputedStyle(e);
}
function We(e) {
  return H(e) ? {
    scrollLeft: e.scrollLeft,
    scrollTop: e.scrollTop
  } : {
    scrollLeft: e.scrollX,
    scrollTop: e.scrollY
  };
}
function te(e) {
  if (ve(e) === "html")
    return e;
  const t = (
    // Step into the shadow DOM of the parent of a slotted node.
    e.assignedSlot || // DOM Element detected.
    e.parentNode || // ShadowRoot detected.
    wt(e) && e.host || // Fallback.
    Q(e)
  );
  return wt(t) ? t.host : t;
}
function zt(e) {
  const t = te(e);
  return de(t) ? e.ownerDocument ? e.ownerDocument.body : e.body : G(t) && Ee(t) ? t : zt(t);
}
function Je(e, t, n) {
  var r;
  t === void 0 && (t = []), n === void 0 && (n = !0);
  const i = zt(e), o = i === ((r = e.ownerDocument) == null ? void 0 : r.body), s = F(i);
  if (o) {
    const a = qe(s);
    return t.concat(s, s.visualViewport || [], Ee(i) ? i : [], a && n ? Je(a) : []);
  }
  return t.concat(i, Je(i, [], n));
}
function qe(e) {
  return e.parent && Object.getPrototypeOf(e.parent) ? e.frameElement : null;
}
function Yt(e) {
  const t = z(e);
  let n = parseFloat(t.width) || 0, r = parseFloat(t.height) || 0;
  const i = G(e), o = i ? e.offsetWidth : n, s = i ? e.offsetHeight : r, a = ke(n) !== o || ke(r) !== s;
  return a && (n = o, r = s), {
    width: n,
    height: r,
    $: a
  };
}
function Gt(e) {
  return H(e) ? e : e.contextElement;
}
function le(e) {
  const t = Gt(e);
  if (!G(t))
    return ee(1);
  const n = t.getBoundingClientRect(), {
    width: r,
    height: i,
    $: o
  } = Yt(t);
  let s = (o ? ke(n.width) : n.width) / r, a = (o ? ke(n.height) : n.height) / i;
  return (!s || !Number.isFinite(s)) && (s = 1), (!a || !Number.isFinite(a)) && (a = 1), {
    x: s,
    y: a
  };
}
const nr = /* @__PURE__ */ ee(0);
function Kt(e) {
  const t = F(e);
  return !at() || !t.visualViewport ? nr : {
    x: t.visualViewport.offsetLeft,
    y: t.visualViewport.offsetTop
  };
}
function rr(e, t, n) {
  return t === void 0 && (t = !1), !n || t && n !== F(e) ? !1 : t;
}
function we(e, t, n, r) {
  t === void 0 && (t = !1), n === void 0 && (n = !1);
  const i = e.getBoundingClientRect(), o = Gt(e);
  let s = ee(1);
  t && (r ? H(r) && (s = le(r)) : s = le(e));
  const a = rr(o, n, r) ? Kt(o) : ee(0);
  let c = (i.left + a.x) / s.x, u = (i.top + a.y) / s.y, l = i.width / s.x, f = i.height / s.y;
  if (o) {
    const h = F(o), d = r && H(r) ? F(r) : r;
    let m = h, p = qe(m);
    for (; p && r && d !== m; ) {
      const g = le(p), v = p.getBoundingClientRect(), S = z(p), k = v.left + (p.clientLeft + parseFloat(S.paddingLeft)) * g.x, _ = v.top + (p.clientTop + parseFloat(S.paddingTop)) * g.y;
      c *= g.x, u *= g.y, l *= g.x, f *= g.y, c += k, u += _, m = F(p), p = qe(m);
    }
  }
  return Le({
    width: l,
    height: f,
    x: c,
    y: u
  });
}
function ir(e) {
  let {
    elements: t,
    rect: n,
    offsetParent: r,
    strategy: i
  } = e;
  const o = i === "fixed", s = Q(r), a = t ? je(t.floating) : !1;
  if (r === s || a && o)
    return n;
  let c = {
    scrollLeft: 0,
    scrollTop: 0
  }, u = ee(1);
  const l = ee(0), f = G(r);
  if ((f || !f && !o) && ((ve(r) !== "body" || Ee(s)) && (c = We(r)), G(r))) {
    const h = we(r);
    u = le(r), l.x = h.x + r.clientLeft, l.y = h.y + r.clientTop;
  }
  return {
    width: n.width * u.x,
    height: n.height * u.y,
    x: n.x * u.x - c.scrollLeft * u.x + l.x,
    y: n.y * u.y - c.scrollTop * u.y + l.y
  };
}
function or(e) {
  return Array.from(e.getClientRects());
}
function Xe(e, t) {
  const n = We(e).scrollLeft;
  return t ? t.left + n : we(Q(e)).left + n;
}
function sr(e) {
  const t = Q(e), n = We(e), r = e.ownerDocument.body, i = ue(t.scrollWidth, t.clientWidth, r.scrollWidth, r.clientWidth), o = ue(t.scrollHeight, t.clientHeight, r.scrollHeight, r.clientHeight);
  let s = -n.scrollLeft + Xe(e);
  const a = -n.scrollTop;
  return z(r).direction === "rtl" && (s += ue(t.clientWidth, r.clientWidth) - i), {
    width: i,
    height: o,
    x: s,
    y: a
  };
}
function ar(e, t) {
  const n = F(e), r = Q(e), i = n.visualViewport;
  let o = r.clientWidth, s = r.clientHeight, a = 0, c = 0;
  if (i) {
    o = i.width, s = i.height;
    const u = at();
    (!u || u && t === "fixed") && (a = i.offsetLeft, c = i.offsetTop);
  }
  return {
    width: o,
    height: s,
    x: a,
    y: c
  };
}
function cr(e, t) {
  const n = we(e, !0, t === "fixed"), r = n.top + e.clientTop, i = n.left + e.clientLeft, o = G(e) ? le(e) : ee(1), s = e.clientWidth * o.x, a = e.clientHeight * o.y, c = i * o.x, u = r * o.y;
  return {
    width: s,
    height: a,
    x: c,
    y: u
  };
}
function St(e, t, n) {
  let r;
  if (t === "viewport")
    r = ar(e, n);
  else if (t === "document")
    r = sr(Q(e));
  else if (H(t))
    r = cr(t, n);
  else {
    const i = Kt(e);
    r = {
      ...t,
      x: t.x - i.x,
      y: t.y - i.y
    };
  }
  return Le(r);
}
function Qt(e, t) {
  const n = te(e);
  return n === t || !H(n) || de(n) ? !1 : z(n).position === "fixed" || Qt(n, t);
}
function ur(e, t) {
  const n = t.get(e);
  if (n)
    return n;
  let r = Je(e, [], !1).filter((a) => H(a) && ve(a) !== "body"), i = null;
  const o = z(e).position === "fixed";
  let s = o ? te(e) : e;
  for (; H(s) && !de(s); ) {
    const a = z(s), c = st(s);
    !c && a.position === "fixed" && (i = null), (o ? !c && !i : !c && a.position === "static" && !!i && ["absolute", "fixed"].includes(i.position) || Ee(s) && !c && Qt(e, s)) ? r = r.filter((l) => l !== s) : i = a, s = te(s);
  }
  return t.set(e, r), r;
}
function lr(e) {
  let {
    element: t,
    boundary: n,
    rootBoundary: r,
    strategy: i
  } = e;
  const s = [...n === "clippingAncestors" ? je(t) ? [] : ur(t, this._c) : [].concat(n), r], a = s[0], c = s.reduce((u, l) => {
    const f = St(t, l, i);
    return u.top = ue(f.top, u.top), u.right = Ke(f.right, u.right), u.bottom = Ke(f.bottom, u.bottom), u.left = ue(f.left, u.left), u;
  }, St(t, a, i));
  return {
    width: c.right - c.left,
    height: c.bottom - c.top,
    x: c.left,
    y: c.top
  };
}
function fr(e) {
  const {
    width: t,
    height: n
  } = Yt(e);
  return {
    width: t,
    height: n
  };
}
function dr(e, t, n) {
  const r = G(t), i = Q(t), o = n === "fixed", s = we(e, !0, o, t);
  let a = {
    scrollLeft: 0,
    scrollTop: 0
  };
  const c = ee(0);
  if (r || !r && !o)
    if ((ve(t) !== "body" || Ee(i)) && (a = We(t)), r) {
      const d = we(t, !0, o, t);
      c.x = d.x + t.clientLeft, c.y = d.y + t.clientTop;
    } else i && (c.x = Xe(i));
  let u = 0, l = 0;
  if (i && !r && !o) {
    const d = i.getBoundingClientRect();
    l = d.top + a.scrollTop, u = d.left + a.scrollLeft - // RTL <body> scrollbar.
    Xe(i, d);
  }
  const f = s.left + a.scrollLeft - c.x - u, h = s.top + a.scrollTop - c.y - l;
  return {
    x: f,
    y: h,
    width: s.width,
    height: s.height
  };
}
function He(e) {
  return z(e).position === "static";
}
function xt(e, t) {
  if (!G(e) || z(e).position === "fixed")
    return null;
  if (t)
    return t(e);
  let n = e.offsetParent;
  return Q(e) === n && (n = n.ownerDocument.body), n;
}
function Jt(e, t) {
  const n = F(e);
  if (je(e))
    return n;
  if (!G(e)) {
    let i = te(e);
    for (; i && !de(i); ) {
      if (H(i) && !He(i))
        return i;
      i = te(i);
    }
    return n;
  }
  let r = xt(e, t);
  for (; r && er(r) && He(r); )
    r = xt(r, t);
  return r && de(r) && He(r) && !st(r) ? n : r || tr(e) || n;
}
const pr = async function(e) {
  const t = this.getOffsetParent || Jt, n = this.getDimensions, r = await n(e.floating);
  return {
    reference: dr(e.reference, await t(e.floating), e.strategy),
    floating: {
      x: 0,
      y: 0,
      width: r.width,
      height: r.height
    }
  };
};
function hr(e) {
  return z(e).direction === "rtl";
}
const vr = {
  convertOffsetParentRelativeRectToViewportRelativeRect: ir,
  getDocumentElement: Q,
  getClippingRect: lr,
  getOffsetParent: Jt,
  getElementRects: pr,
  getClientRects: or,
  getDimensions: fr,
  getScale: le,
  isElement: H,
  isRTL: hr
}, mr = Xn, gr = Zn, br = Jn, yr = (e, t, n) => {
  const r = /* @__PURE__ */ new Map(), i = {
    platform: vr,
    ...n
  }, o = {
    ...i.platform,
    _c: r
  };
  return Qn(e, t, {
    ...i,
    platform: o
  });
};
var wr = typeof globalThis < "u" ? globalThis : typeof window < "u" ? window : typeof global < "u" ? global : typeof self < "u" ? self : {};
function Sr(e) {
  return e && e.__esModule && Object.prototype.hasOwnProperty.call(e, "default") ? e.default : e;
}
var qt = { exports: {} };
(function(e) {
  (function(t, n) {
    e.exports ? e.exports = n() : t.log = n();
  })(wr, function() {
    var t = function() {
    }, n = "undefined", r = typeof window !== n && typeof window.navigator !== n && /Trident\/|MSIE /.test(window.navigator.userAgent), i = [
      "trace",
      "debug",
      "info",
      "warn",
      "error"
    ], o = {}, s = null;
    function a(p, g) {
      var v = p[g];
      if (typeof v.bind == "function")
        return v.bind(p);
      try {
        return Function.prototype.bind.call(v, p);
      } catch {
        return function() {
          return Function.prototype.apply.apply(v, [p, arguments]);
        };
      }
    }
    function c() {
      console.log && (console.log.apply ? console.log.apply(console, arguments) : Function.prototype.apply.apply(console.log, [console, arguments])), console.trace && console.trace();
    }
    function u(p) {
      return p === "debug" && (p = "log"), typeof console === n ? !1 : p === "trace" && r ? c : console[p] !== void 0 ? a(console, p) : console.log !== void 0 ? a(console, "log") : t;
    }
    function l() {
      for (var p = this.getLevel(), g = 0; g < i.length; g++) {
        var v = i[g];
        this[v] = g < p ? t : this.methodFactory(v, p, this.name);
      }
      if (this.log = this.debug, typeof console === n && p < this.levels.SILENT)
        return "No console available for logging";
    }
    function f(p) {
      return function() {
        typeof console !== n && (l.call(this), this[p].apply(this, arguments));
      };
    }
    function h(p, g, v) {
      return u(p) || f.apply(this, arguments);
    }
    function d(p, g) {
      var v = this, S, k, _, E = "loglevel";
      typeof p == "string" ? E += ":" + p : typeof p == "symbol" && (E = void 0);
      function re(w) {
        var x = (i[w] || "silent").toUpperCase();
        if (!(typeof window === n || !E)) {
          try {
            window.localStorage[E] = x;
            return;
          } catch {
          }
          try {
            window.document.cookie = encodeURIComponent(E) + "=" + x + ";";
          } catch {
          }
        }
      }
      function C() {
        var w;
        if (!(typeof window === n || !E)) {
          try {
            w = window.localStorage[E];
          } catch {
          }
          if (typeof w === n)
            try {
              var x = window.document.cookie, ie = encodeURIComponent(E), q = x.indexOf(ie + "=");
              q !== -1 && (w = /^([^;]+)/.exec(
                x.slice(q + ie.length + 1)
              )[1]);
            } catch {
            }
          return v.levels[w] === void 0 && (w = void 0), w;
        }
      }
      function T() {
        if (!(typeof window === n || !E)) {
          try {
            window.localStorage.removeItem(E);
          } catch {
          }
          try {
            window.document.cookie = encodeURIComponent(E) + "=; expires=Thu, 01 Jan 1970 00:00:00 UTC";
          } catch {
          }
        }
      }
      function I(w) {
        var x = w;
        if (typeof x == "string" && v.levels[x.toUpperCase()] !== void 0 && (x = v.levels[x.toUpperCase()]), typeof x == "number" && x >= 0 && x <= v.levels.SILENT)
          return x;
        throw new TypeError("log.setLevel() called with invalid level: " + w);
      }
      v.name = p, v.levels = {
        TRACE: 0,
        DEBUG: 1,
        INFO: 2,
        WARN: 3,
        ERROR: 4,
        SILENT: 5
      }, v.methodFactory = g || h, v.getLevel = function() {
        return _ ?? k ?? S;
      }, v.setLevel = function(w, x) {
        return _ = I(w), x !== !1 && re(_), l.call(v);
      }, v.setDefaultLevel = function(w) {
        k = I(w), C() || v.setLevel(w, !1);
      }, v.resetLevel = function() {
        _ = null, T(), l.call(v);
      }, v.enableAll = function(w) {
        v.setLevel(v.levels.TRACE, w);
      }, v.disableAll = function(w) {
        v.setLevel(v.levels.SILENT, w);
      }, v.rebuild = function() {
        if (s !== v && (S = I(s.getLevel())), l.call(v), s === v)
          for (var w in o)
            o[w].rebuild();
      }, S = I(
        s ? s.getLevel() : "WARN"
      );
      var N = C();
      N != null && (_ = I(N)), l.call(v);
    }
    s = new d(), s.getLogger = function(g) {
      if (typeof g != "symbol" && typeof g != "string" || g === "")
        throw new TypeError("You must supply a name when creating a logger.");
      var v = o[g];
      return v || (v = o[g] = new d(
        g,
        s.methodFactory
      )), v;
    };
    var m = typeof window !== n ? window.log : void 0;
    return s.noConflict = function() {
      return typeof window !== n && window.log === s && (window.log = m), s;
    }, s.getLoggers = function() {
      return o;
    }, s.default = s, s;
  });
})(qt);
var xr = qt.exports;
const Er = /* @__PURE__ */ Sr(xr);
var Ze = function(e, t) {
  return Ze = Object.setPrototypeOf || { __proto__: [] } instanceof Array && function(n, r) {
    n.__proto__ = r;
  } || function(n, r) {
    for (var i in r) Object.prototype.hasOwnProperty.call(r, i) && (n[i] = r[i]);
  }, Ze(e, t);
};
function J(e, t) {
  if (typeof t != "function" && t !== null)
    throw new TypeError("Class extends value " + String(t) + " is not a constructor or null");
  Ze(e, t);
  function n() {
    this.constructor = e;
  }
  e.prototype = t === null ? Object.create(t) : (n.prototype = t.prototype, new n());
}
function Tr(e, t, n, r) {
  function i(o) {
    return o instanceof n ? o : new n(function(s) {
      s(o);
    });
  }
  return new (n || (n = Promise))(function(o, s) {
    function a(l) {
      try {
        u(r.next(l));
      } catch (f) {
        s(f);
      }
    }
    function c(l) {
      try {
        u(r.throw(l));
      } catch (f) {
        s(f);
      }
    }
    function u(l) {
      l.done ? o(l.value) : i(l.value).then(a, c);
    }
    u((r = r.apply(e, t || [])).next());
  });
}
function Xt(e, t) {
  var n = { label: 0, sent: function() {
    if (o[0] & 1) throw o[1];
    return o[1];
  }, trys: [], ops: [] }, r, i, o, s = Object.create((typeof Iterator == "function" ? Iterator : Object).prototype);
  return s.next = a(0), s.throw = a(1), s.return = a(2), typeof Symbol == "function" && (s[Symbol.iterator] = function() {
    return this;
  }), s;
  function a(u) {
    return function(l) {
      return c([u, l]);
    };
  }
  function c(u) {
    if (r) throw new TypeError("Generator is already executing.");
    for (; s && (s = 0, u[0] && (n = 0)), n; ) try {
      if (r = 1, i && (o = u[0] & 2 ? i.return : u[0] ? i.throw || ((o = i.return) && o.call(i), 0) : i.next) && !(o = o.call(i, u[1])).done) return o;
      switch (i = 0, o && (u = [u[0] & 2, o.value]), u[0]) {
        case 0:
        case 1:
          o = u;
          break;
        case 4:
          return n.label++, { value: u[1], done: !1 };
        case 5:
          n.label++, i = u[1], u = [0];
          continue;
        case 7:
          u = n.ops.pop(), n.trys.pop();
          continue;
        default:
          if (o = n.trys, !(o = o.length > 0 && o[o.length - 1]) && (u[0] === 6 || u[0] === 2)) {
            n = 0;
            continue;
          }
          if (u[0] === 3 && (!o || u[1] > o[0] && u[1] < o[3])) {
            n.label = u[1];
            break;
          }
          if (u[0] === 6 && n.label < o[1]) {
            n.label = o[1], o = u;
            break;
          }
          if (o && n.label < o[2]) {
            n.label = o[2], n.ops.push(u);
            break;
          }
          o[2] && n.ops.pop(), n.trys.pop();
          continue;
      }
      u = t.call(e, n);
    } catch (l) {
      u = [6, l], i = 0;
    } finally {
      r = o = 0;
    }
    if (u[0] & 5) throw u[1];
    return { value: u[0] ? u[1] : void 0, done: !0 };
  }
}
function pe(e) {
  var t = typeof Symbol == "function" && Symbol.iterator, n = t && e[t], r = 0;
  if (n) return n.call(e);
  if (e && typeof e.length == "number") return {
    next: function() {
      return e && r >= e.length && (e = void 0), { value: e && e[r++], done: !e };
    }
  };
  throw new TypeError(t ? "Object is not iterable." : "Symbol.iterator is not defined.");
}
function he(e, t) {
  var n = typeof Symbol == "function" && e[Symbol.iterator];
  if (!n) return e;
  var r = n.call(e), i, o = [], s;
  try {
    for (; (t === void 0 || t-- > 0) && !(i = r.next()).done; ) o.push(i.value);
  } catch (a) {
    s = { error: a };
  } finally {
    try {
      i && !i.done && (n = r.return) && n.call(r);
    } finally {
      if (s) throw s.error;
    }
  }
  return o;
}
function Se(e, t, n) {
  if (n || arguments.length === 2) for (var r = 0, i = t.length, o; r < i; r++)
    (o || !(r in t)) && (o || (o = Array.prototype.slice.call(t, 0, r)), o[r] = t[r]);
  return e.concat(o || Array.prototype.slice.call(t));
}
function fe(e) {
  return this instanceof fe ? (this.v = e, this) : new fe(e);
}
function Cr(e, t, n) {
  if (!Symbol.asyncIterator) throw new TypeError("Symbol.asyncIterator is not defined.");
  var r = n.apply(e, t || []), i, o = [];
  return i = Object.create((typeof AsyncIterator == "function" ? AsyncIterator : Object).prototype), a("next"), a("throw"), a("return", s), i[Symbol.asyncIterator] = function() {
    return this;
  }, i;
  function s(d) {
    return function(m) {
      return Promise.resolve(m).then(d, f);
    };
  }
  function a(d, m) {
    r[d] && (i[d] = function(p) {
      return new Promise(function(g, v) {
        o.push([d, p, g, v]) > 1 || c(d, p);
      });
    }, m && (i[d] = m(i[d])));
  }
  function c(d, m) {
    try {
      u(r[d](m));
    } catch (p) {
      h(o[0][3], p);
    }
  }
  function u(d) {
    d.value instanceof fe ? Promise.resolve(d.value.v).then(l, f) : h(o[0][2], d);
  }
  function l(d) {
    c("next", d);
  }
  function f(d) {
    c("throw", d);
  }
  function h(d, m) {
    d(m), o.shift(), o.length && c(o[0][0], o[0][1]);
  }
}
function Pr(e) {
  if (!Symbol.asyncIterator) throw new TypeError("Symbol.asyncIterator is not defined.");
  var t = e[Symbol.asyncIterator], n;
  return t ? t.call(e) : (e = typeof pe == "function" ? pe(e) : e[Symbol.iterator](), n = {}, r("next"), r("throw"), r("return"), n[Symbol.asyncIterator] = function() {
    return this;
  }, n);
  function r(o) {
    n[o] = e[o] && function(s) {
      return new Promise(function(a, c) {
        s = e[o](s), i(a, c, s.done, s.value);
      });
    };
  }
  function i(o, s, a, c) {
    Promise.resolve(c).then(function(u) {
      o({ value: u, done: a });
    }, s);
  }
}
function P(e) {
  return typeof e == "function";
}
function ct(e) {
  var t = function(r) {
    Error.call(r), r.stack = new Error().stack;
  }, n = e(t);
  return n.prototype = Object.create(Error.prototype), n.prototype.constructor = n, n;
}
var ze = ct(function(e) {
  return function(n) {
    e(this), this.message = n ? n.length + ` errors occurred during unsubscription:
` + n.map(function(r, i) {
      return i + 1 + ") " + r.toString();
    }).join(`
  `) : "", this.name = "UnsubscriptionError", this.errors = n;
  };
});
function _e(e, t) {
  if (e) {
    var n = e.indexOf(t);
    0 <= n && e.splice(n, 1);
  }
}
var Te = function() {
  function e(t) {
    this.initialTeardown = t, this.closed = !1, this._parentage = null, this._finalizers = null;
  }
  return e.prototype.unsubscribe = function() {
    var t, n, r, i, o;
    if (!this.closed) {
      this.closed = !0;
      var s = this._parentage;
      if (s)
        if (this._parentage = null, Array.isArray(s))
          try {
            for (var a = pe(s), c = a.next(); !c.done; c = a.next()) {
              var u = c.value;
              u.remove(this);
            }
          } catch (p) {
            t = { error: p };
          } finally {
            try {
              c && !c.done && (n = a.return) && n.call(a);
            } finally {
              if (t) throw t.error;
            }
          }
        else
          s.remove(this);
      var l = this.initialTeardown;
      if (P(l))
        try {
          l();
        } catch (p) {
          o = p instanceof ze ? p.errors : [p];
        }
      var f = this._finalizers;
      if (f) {
        this._finalizers = null;
        try {
          for (var h = pe(f), d = h.next(); !d.done; d = h.next()) {
            var m = d.value;
            try {
              Et(m);
            } catch (p) {
              o = o ?? [], p instanceof ze ? o = Se(Se([], he(o)), he(p.errors)) : o.push(p);
            }
          }
        } catch (p) {
          r = { error: p };
        } finally {
          try {
            d && !d.done && (i = h.return) && i.call(h);
          } finally {
            if (r) throw r.error;
          }
        }
      }
      if (o)
        throw new ze(o);
    }
  }, e.prototype.add = function(t) {
    var n;
    if (t && t !== this)
      if (this.closed)
        Et(t);
      else {
        if (t instanceof e) {
          if (t.closed || t._hasParent(this))
            return;
          t._addParent(this);
        }
        (this._finalizers = (n = this._finalizers) !== null && n !== void 0 ? n : []).push(t);
      }
  }, e.prototype._hasParent = function(t) {
    var n = this._parentage;
    return n === t || Array.isArray(n) && n.includes(t);
  }, e.prototype._addParent = function(t) {
    var n = this._parentage;
    this._parentage = Array.isArray(n) ? (n.push(t), n) : n ? [n, t] : t;
  }, e.prototype._removeParent = function(t) {
    var n = this._parentage;
    n === t ? this._parentage = null : Array.isArray(n) && _e(n, t);
  }, e.prototype.remove = function(t) {
    var n = this._finalizers;
    n && _e(n, t), t instanceof e && t._removeParent(this);
  }, e.EMPTY = function() {
    var t = new e();
    return t.closed = !0, t;
  }(), e;
}(), Zt = Te.EMPTY;
function en(e) {
  return e instanceof Te || e && "closed" in e && P(e.remove) && P(e.add) && P(e.unsubscribe);
}
function Et(e) {
  P(e) ? e() : e.unsubscribe();
}
var tn = {
  onUnhandledError: null,
  onStoppedNotification: null,
  Promise: void 0,
  useDeprecatedSynchronousErrorHandling: !1,
  useDeprecatedNextContext: !1
}, nn = {
  setTimeout: function(e, t) {
    for (var n = [], r = 2; r < arguments.length; r++)
      n[r - 2] = arguments[r];
    return setTimeout.apply(void 0, Se([e, t], he(n)));
  },
  clearTimeout: function(e) {
    var t = nn.delegate;
    return ((t == null ? void 0 : t.clearTimeout) || clearTimeout)(e);
  },
  delegate: void 0
};
function rn(e) {
  nn.setTimeout(function() {
    throw e;
  });
}
function Ie() {
}
function Ae(e) {
  e();
}
var ut = function(e) {
  J(t, e);
  function t(n) {
    var r = e.call(this) || this;
    return r.isStopped = !1, n ? (r.destination = n, en(n) && n.add(r)) : r.destination = Lr, r;
  }
  return t.create = function(n, r, i) {
    return new et(n, r, i);
  }, t.prototype.next = function(n) {
    this.isStopped || this._next(n);
  }, t.prototype.error = function(n) {
    this.isStopped || (this.isStopped = !0, this._error(n));
  }, t.prototype.complete = function() {
    this.isStopped || (this.isStopped = !0, this._complete());
  }, t.prototype.unsubscribe = function() {
    this.closed || (this.isStopped = !0, e.prototype.unsubscribe.call(this), this.destination = null);
  }, t.prototype._next = function(n) {
    this.destination.next(n);
  }, t.prototype._error = function(n) {
    try {
      this.destination.error(n);
    } finally {
      this.unsubscribe();
    }
  }, t.prototype._complete = function() {
    try {
      this.destination.complete();
    } finally {
      this.unsubscribe();
    }
  }, t;
}(Te), Ar = Function.prototype.bind;
function Ye(e, t) {
  return Ar.call(e, t);
}
var kr = function() {
  function e(t) {
    this.partialObserver = t;
  }
  return e.prototype.next = function(t) {
    var n = this.partialObserver;
    if (n.next)
      try {
        n.next(t);
      } catch (r) {
        Ce(r);
      }
  }, e.prototype.error = function(t) {
    var n = this.partialObserver;
    if (n.error)
      try {
        n.error(t);
      } catch (r) {
        Ce(r);
      }
    else
      Ce(t);
  }, e.prototype.complete = function() {
    var t = this.partialObserver;
    if (t.complete)
      try {
        t.complete();
      } catch (n) {
        Ce(n);
      }
  }, e;
}(), et = function(e) {
  J(t, e);
  function t(n, r, i) {
    var o = e.call(this) || this, s;
    if (P(n) || !n)
      s = {
        next: n ?? void 0,
        error: r ?? void 0,
        complete: i ?? void 0
      };
    else {
      var a;
      o && tn.useDeprecatedNextContext ? (a = Object.create(n), a.unsubscribe = function() {
        return o.unsubscribe();
      }, s = {
        next: n.next && Ye(n.next, a),
        error: n.error && Ye(n.error, a),
        complete: n.complete && Ye(n.complete, a)
      }) : s = n;
    }
    return o.destination = new kr(s), o;
  }
  return t;
}(ut);
function Ce(e) {
  rn(e);
}
function Or(e) {
  throw e;
}
var Lr = {
  closed: !0,
  next: Ie,
  error: Or,
  complete: Ie
}, lt = function() {
  return typeof Symbol == "function" && Symbol.observable || "@@observable";
}();
function ft(e) {
  return e;
}
function _r(e) {
  return e.length === 0 ? ft : e.length === 1 ? e[0] : function(n) {
    return e.reduce(function(r, i) {
      return i(r);
    }, n);
  };
}
var O = function() {
  function e(t) {
    t && (this._subscribe = t);
  }
  return e.prototype.lift = function(t) {
    var n = new e();
    return n.source = this, n.operator = t, n;
  }, e.prototype.subscribe = function(t, n, r) {
    var i = this, o = Dr(t) ? t : new et(t, n, r);
    return Ae(function() {
      var s = i, a = s.operator, c = s.source;
      o.add(a ? a.call(o, c) : c ? i._subscribe(o) : i._trySubscribe(o));
    }), o;
  }, e.prototype._trySubscribe = function(t) {
    try {
      return this._subscribe(t);
    } catch (n) {
      t.error(n);
    }
  }, e.prototype.forEach = function(t, n) {
    var r = this;
    return n = Tt(n), new n(function(i, o) {
      var s = new et({
        next: function(a) {
          try {
            t(a);
          } catch (c) {
            o(c), s.unsubscribe();
          }
        },
        error: o,
        complete: i
      });
      r.subscribe(s);
    });
  }, e.prototype._subscribe = function(t) {
    var n;
    return (n = this.source) === null || n === void 0 ? void 0 : n.subscribe(t);
  }, e.prototype[lt] = function() {
    return this;
  }, e.prototype.pipe = function() {
    for (var t = [], n = 0; n < arguments.length; n++)
      t[n] = arguments[n];
    return _r(t)(this);
  }, e.prototype.toPromise = function(t) {
    var n = this;
    return t = Tt(t), new t(function(r, i) {
      var o;
      n.subscribe(function(s) {
        return o = s;
      }, function(s) {
        return i(s);
      }, function() {
        return r(o);
      });
    });
  }, e.create = function(t) {
    return new e(t);
  }, e;
}();
function Tt(e) {
  var t;
  return (t = e ?? tn.Promise) !== null && t !== void 0 ? t : Promise;
}
function Ir(e) {
  return e && P(e.next) && P(e.error) && P(e.complete);
}
function Dr(e) {
  return e && e instanceof ut || Ir(e) && en(e);
}
function Mr(e) {
  return P(e == null ? void 0 : e.lift);
}
function V(e) {
  return function(t) {
    if (Mr(t))
      return t.lift(function(n) {
        try {
          return e(n, this);
        } catch (r) {
          this.error(r);
        }
      });
    throw new TypeError("Unable to lift unknown Observable type");
  };
}
function W(e, t, n, r, i) {
  return new Rr(e, t, n, r, i);
}
var Rr = function(e) {
  J(t, e);
  function t(n, r, i, o, s, a) {
    var c = e.call(this, n) || this;
    return c.onFinalize = s, c.shouldUnsubscribe = a, c._next = r ? function(u) {
      try {
        r(u);
      } catch (l) {
        n.error(l);
      }
    } : e.prototype._next, c._error = o ? function(u) {
      try {
        o(u);
      } catch (l) {
        n.error(l);
      } finally {
        this.unsubscribe();
      }
    } : e.prototype._error, c._complete = i ? function() {
      try {
        i();
      } catch (u) {
        n.error(u);
      } finally {
        this.unsubscribe();
      }
    } : e.prototype._complete, c;
  }
  return t.prototype.unsubscribe = function() {
    var n;
    if (!this.shouldUnsubscribe || this.shouldUnsubscribe()) {
      var r = this.closed;
      e.prototype.unsubscribe.call(this), !r && ((n = this.onFinalize) === null || n === void 0 || n.call(this));
    }
  }, t;
}(ut), Nr = ct(function(e) {
  return function() {
    e(this), this.name = "ObjectUnsubscribedError", this.message = "object unsubscribed";
  };
}), ne = function(e) {
  J(t, e);
  function t() {
    var n = e.call(this) || this;
    return n.closed = !1, n.currentObservers = null, n.observers = [], n.isStopped = !1, n.hasError = !1, n.thrownError = null, n;
  }
  return t.prototype.lift = function(n) {
    var r = new Ct(this, this);
    return r.operator = n, r;
  }, t.prototype._throwIfClosed = function() {
    if (this.closed)
      throw new Nr();
  }, t.prototype.next = function(n) {
    var r = this;
    Ae(function() {
      var i, o;
      if (r._throwIfClosed(), !r.isStopped) {
        r.currentObservers || (r.currentObservers = Array.from(r.observers));
        try {
          for (var s = pe(r.currentObservers), a = s.next(); !a.done; a = s.next()) {
            var c = a.value;
            c.next(n);
          }
        } catch (u) {
          i = { error: u };
        } finally {
          try {
            a && !a.done && (o = s.return) && o.call(s);
          } finally {
            if (i) throw i.error;
          }
        }
      }
    });
  }, t.prototype.error = function(n) {
    var r = this;
    Ae(function() {
      if (r._throwIfClosed(), !r.isStopped) {
        r.hasError = r.isStopped = !0, r.thrownError = n;
        for (var i = r.observers; i.length; )
          i.shift().error(n);
      }
    });
  }, t.prototype.complete = function() {
    var n = this;
    Ae(function() {
      if (n._throwIfClosed(), !n.isStopped) {
        n.isStopped = !0;
        for (var r = n.observers; r.length; )
          r.shift().complete();
      }
    });
  }, t.prototype.unsubscribe = function() {
    this.isStopped = this.closed = !0, this.observers = this.currentObservers = null;
  }, Object.defineProperty(t.prototype, "observed", {
    get: function() {
      var n;
      return ((n = this.observers) === null || n === void 0 ? void 0 : n.length) > 0;
    },
    enumerable: !1,
    configurable: !0
  }), t.prototype._trySubscribe = function(n) {
    return this._throwIfClosed(), e.prototype._trySubscribe.call(this, n);
  }, t.prototype._subscribe = function(n) {
    return this._throwIfClosed(), this._checkFinalizedStatuses(n), this._innerSubscribe(n);
  }, t.prototype._innerSubscribe = function(n) {
    var r = this, i = this, o = i.hasError, s = i.isStopped, a = i.observers;
    return o || s ? Zt : (this.currentObservers = null, a.push(n), new Te(function() {
      r.currentObservers = null, _e(a, n);
    }));
  }, t.prototype._checkFinalizedStatuses = function(n) {
    var r = this, i = r.hasError, o = r.thrownError, s = r.isStopped;
    i ? n.error(o) : s && n.complete();
  }, t.prototype.asObservable = function() {
    var n = new O();
    return n.source = this, n;
  }, t.create = function(n, r) {
    return new Ct(n, r);
  }, t;
}(O), Ct = function(e) {
  J(t, e);
  function t(n, r) {
    var i = e.call(this) || this;
    return i.destination = n, i.source = r, i;
  }
  return t.prototype.next = function(n) {
    var r, i;
    (i = (r = this.destination) === null || r === void 0 ? void 0 : r.next) === null || i === void 0 || i.call(r, n);
  }, t.prototype.error = function(n) {
    var r, i;
    (i = (r = this.destination) === null || r === void 0 ? void 0 : r.error) === null || i === void 0 || i.call(r, n);
  }, t.prototype.complete = function() {
    var n, r;
    (r = (n = this.destination) === null || n === void 0 ? void 0 : n.complete) === null || r === void 0 || r.call(n);
  }, t.prototype._subscribe = function(n) {
    var r, i;
    return (i = (r = this.source) === null || r === void 0 ? void 0 : r.subscribe(n)) !== null && i !== void 0 ? i : Zt;
  }, t;
}(ne), on = function(e) {
  J(t, e);
  function t(n) {
    var r = e.call(this) || this;
    return r._value = n, r;
  }
  return Object.defineProperty(t.prototype, "value", {
    get: function() {
      return this.getValue();
    },
    enumerable: !1,
    configurable: !0
  }), t.prototype._subscribe = function(n) {
    var r = e.prototype._subscribe.call(this, n);
    return !r.closed && n.next(this._value), r;
  }, t.prototype.getValue = function() {
    var n = this, r = n.hasError, i = n.thrownError, o = n._value;
    if (r)
      throw i;
    return this._throwIfClosed(), o;
  }, t.prototype.next = function(n) {
    e.prototype.next.call(this, this._value = n);
  }, t;
}(ne), $r = {
  now: function() {
    return Date.now();
  },
  delegate: void 0
}, Ur = function(e) {
  J(t, e);
  function t(n, r) {
    return e.call(this) || this;
  }
  return t.prototype.schedule = function(n, r) {
    return this;
  }, t;
}(Te), Pt = {
  setInterval: function(e, t) {
    for (var n = [], r = 2; r < arguments.length; r++)
      n[r - 2] = arguments[r];
    return setInterval.apply(void 0, Se([e, t], he(n)));
  },
  clearInterval: function(e) {
    return clearInterval(e);
  },
  delegate: void 0
}, Fr = function(e) {
  J(t, e);
  function t(n, r) {
    var i = e.call(this, n, r) || this;
    return i.scheduler = n, i.work = r, i.pending = !1, i;
  }
  return t.prototype.schedule = function(n, r) {
    var i;
    if (r === void 0 && (r = 0), this.closed)
      return this;
    this.state = n;
    var o = this.id, s = this.scheduler;
    return o != null && (this.id = this.recycleAsyncId(s, o, r)), this.pending = !0, this.delay = r, this.id = (i = this.id) !== null && i !== void 0 ? i : this.requestAsyncId(s, this.id, r), this;
  }, t.prototype.requestAsyncId = function(n, r, i) {
    return i === void 0 && (i = 0), Pt.setInterval(n.flush.bind(n, this), i);
  }, t.prototype.recycleAsyncId = function(n, r, i) {
    if (i === void 0 && (i = 0), i != null && this.delay === i && this.pending === !1)
      return r;
    r != null && Pt.clearInterval(r);
  }, t.prototype.execute = function(n, r) {
    if (this.closed)
      return new Error("executing a cancelled action");
    this.pending = !1;
    var i = this._execute(n, r);
    if (i)
      return i;
    this.pending === !1 && this.id != null && (this.id = this.recycleAsyncId(this.scheduler, this.id, null));
  }, t.prototype._execute = function(n, r) {
    var i = !1, o;
    try {
      this.work(n);
    } catch (s) {
      i = !0, o = s || new Error("Scheduled action threw falsy error");
    }
    if (i)
      return this.unsubscribe(), o;
  }, t.prototype.unsubscribe = function() {
    if (!this.closed) {
      var n = this, r = n.id, i = n.scheduler, o = i.actions;
      this.work = this.state = this.scheduler = null, this.pending = !1, _e(o, this), r != null && (this.id = this.recycleAsyncId(i, r, null)), this.delay = null, e.prototype.unsubscribe.call(this);
    }
  }, t;
}(Ur), At = function() {
  function e(t, n) {
    n === void 0 && (n = e.now), this.schedulerActionCtor = t, this.now = n;
  }
  return e.prototype.schedule = function(t, n, r) {
    return n === void 0 && (n = 0), new this.schedulerActionCtor(this, t).schedule(r, n);
  }, e.now = $r.now, e;
}(), jr = function(e) {
  J(t, e);
  function t(n, r) {
    r === void 0 && (r = At.now);
    var i = e.call(this, n, r) || this;
    return i.actions = [], i._active = !1, i;
  }
  return t.prototype.flush = function(n) {
    var r = this.actions;
    if (this._active) {
      r.push(n);
      return;
    }
    var i;
    this._active = !0;
    do
      if (i = n.execute(n.state, n.delay))
        break;
    while (n = r.shift());
    if (this._active = !1, i) {
      for (; n = r.shift(); )
        n.unsubscribe();
      throw i;
    }
  }, t;
}(At), Wr = new jr(Fr), Br = new O(function(e) {
  return e.complete();
});
function Vr(e) {
  return e && P(e.schedule);
}
function sn(e) {
  return e[e.length - 1];
}
function Be(e) {
  return Vr(sn(e)) ? e.pop() : void 0;
}
function Hr(e, t) {
  return typeof sn(e) == "number" ? e.pop() : t;
}
var dt = function(e) {
  return e && typeof e.length == "number" && typeof e != "function";
};
function an(e) {
  return P(e == null ? void 0 : e.then);
}
function cn(e) {
  return P(e[lt]);
}
function un(e) {
  return Symbol.asyncIterator && P(e == null ? void 0 : e[Symbol.asyncIterator]);
}
function ln(e) {
  return new TypeError("You provided " + (e !== null && typeof e == "object" ? "an invalid object" : "'" + e + "'") + " where a stream was expected. You can provide an Observable, Promise, ReadableStream, Array, AsyncIterable, or Iterable.");
}
function zr() {
  return typeof Symbol != "function" || !Symbol.iterator ? "@@iterator" : Symbol.iterator;
}
var fn = zr();
function dn(e) {
  return P(e == null ? void 0 : e[fn]);
}
function pn(e) {
  return Cr(this, arguments, function() {
    var n, r, i, o;
    return Xt(this, function(s) {
      switch (s.label) {
        case 0:
          n = e.getReader(), s.label = 1;
        case 1:
          s.trys.push([1, , 9, 10]), s.label = 2;
        case 2:
          return [4, fe(n.read())];
        case 3:
          return r = s.sent(), i = r.value, o = r.done, o ? [4, fe(void 0)] : [3, 5];
        case 4:
          return [2, s.sent()];
        case 5:
          return [4, fe(i)];
        case 6:
          return [4, s.sent()];
        case 7:
          return s.sent(), [3, 2];
        case 8:
          return [3, 10];
        case 9:
          return n.releaseLock(), [7];
        case 10:
          return [2];
      }
    });
  });
}
function hn(e) {
  return P(e == null ? void 0 : e.getReader);
}
function Y(e) {
  if (e instanceof O)
    return e;
  if (e != null) {
    if (cn(e))
      return Yr(e);
    if (dt(e))
      return Gr(e);
    if (an(e))
      return Kr(e);
    if (un(e))
      return vn(e);
    if (dn(e))
      return Qr(e);
    if (hn(e))
      return Jr(e);
  }
  throw ln(e);
}
function Yr(e) {
  return new O(function(t) {
    var n = e[lt]();
    if (P(n.subscribe))
      return n.subscribe(t);
    throw new TypeError("Provided object does not correctly implement Symbol.observable");
  });
}
function Gr(e) {
  return new O(function(t) {
    for (var n = 0; n < e.length && !t.closed; n++)
      t.next(e[n]);
    t.complete();
  });
}
function Kr(e) {
  return new O(function(t) {
    e.then(function(n) {
      t.closed || (t.next(n), t.complete());
    }, function(n) {
      return t.error(n);
    }).then(null, rn);
  });
}
function Qr(e) {
  return new O(function(t) {
    var n, r;
    try {
      for (var i = pe(e), o = i.next(); !o.done; o = i.next()) {
        var s = o.value;
        if (t.next(s), t.closed)
          return;
      }
    } catch (a) {
      n = { error: a };
    } finally {
      try {
        o && !o.done && (r = i.return) && r.call(i);
      } finally {
        if (n) throw n.error;
      }
    }
    t.complete();
  });
}
function vn(e) {
  return new O(function(t) {
    qr(e, t).catch(function(n) {
      return t.error(n);
    });
  });
}
function Jr(e) {
  return vn(pn(e));
}
function qr(e, t) {
  var n, r, i, o;
  return Tr(this, void 0, void 0, function() {
    var s, a;
    return Xt(this, function(c) {
      switch (c.label) {
        case 0:
          c.trys.push([0, 5, 6, 11]), n = Pr(e), c.label = 1;
        case 1:
          return [4, n.next()];
        case 2:
          if (r = c.sent(), !!r.done) return [3, 4];
          if (s = r.value, t.next(s), t.closed)
            return [2];
          c.label = 3;
        case 3:
          return [3, 1];
        case 4:
          return [3, 11];
        case 5:
          return a = c.sent(), i = { error: a }, [3, 11];
        case 6:
          return c.trys.push([6, , 9, 10]), r && !r.done && (o = n.return) ? [4, o.call(n)] : [3, 8];
        case 7:
          c.sent(), c.label = 8;
        case 8:
          return [3, 10];
        case 9:
          if (i) throw i.error;
          return [7];
        case 10:
          return [7];
        case 11:
          return t.complete(), [2];
      }
    });
  });
}
function Z(e, t, n, r, i) {
  r === void 0 && (r = 0), i === void 0 && (i = !1);
  var o = t.schedule(function() {
    n(), i ? e.add(this.schedule(null, r)) : this.unsubscribe();
  }, r);
  if (e.add(o), !i)
    return o;
}
function mn(e, t) {
  return t === void 0 && (t = 0), V(function(n, r) {
    n.subscribe(W(r, function(i) {
      return Z(r, e, function() {
        return r.next(i);
      }, t);
    }, function() {
      return Z(r, e, function() {
        return r.complete();
      }, t);
    }, function(i) {
      return Z(r, e, function() {
        return r.error(i);
      }, t);
    }));
  });
}
function gn(e, t) {
  return t === void 0 && (t = 0), V(function(n, r) {
    r.add(e.schedule(function() {
      return n.subscribe(r);
    }, t));
  });
}
function Xr(e, t) {
  return Y(e).pipe(gn(t), mn(t));
}
function Zr(e, t) {
  return Y(e).pipe(gn(t), mn(t));
}
function ei(e, t) {
  return new O(function(n) {
    var r = 0;
    return t.schedule(function() {
      r === e.length ? n.complete() : (n.next(e[r++]), n.closed || this.schedule());
    });
  });
}
function ti(e, t) {
  return new O(function(n) {
    var r;
    return Z(n, t, function() {
      r = e[fn](), Z(n, t, function() {
        var i, o, s;
        try {
          i = r.next(), o = i.value, s = i.done;
        } catch (a) {
          n.error(a);
          return;
        }
        s ? n.complete() : n.next(o);
      }, 0, !0);
    }), function() {
      return P(r == null ? void 0 : r.return) && r.return();
    };
  });
}
function bn(e, t) {
  if (!e)
    throw new Error("Iterable cannot be null");
  return new O(function(n) {
    Z(n, t, function() {
      var r = e[Symbol.asyncIterator]();
      Z(n, t, function() {
        r.next().then(function(i) {
          i.done ? n.complete() : n.next(i.value);
        });
      }, 0, !0);
    });
  });
}
function ni(e, t) {
  return bn(pn(e), t);
}
function ri(e, t) {
  if (e != null) {
    if (cn(e))
      return Xr(e, t);
    if (dt(e))
      return ei(e, t);
    if (an(e))
      return Zr(e, t);
    if (un(e))
      return bn(e, t);
    if (dn(e))
      return ti(e, t);
    if (hn(e))
      return ni(e, t);
  }
  throw ln(e);
}
function pt(e, t) {
  return t ? ri(e, t) : Y(e);
}
function kt() {
  for (var e = [], t = 0; t < arguments.length; t++)
    e[t] = arguments[t];
  var n = Be(e);
  return pt(e, n);
}
function ii(e) {
  return e instanceof Date && !isNaN(e);
}
var oi = ct(function(e) {
  return function(n) {
    n === void 0 && (n = null), e(this), this.message = "Timeout has occurred", this.name = "TimeoutError", this.info = n;
  };
});
function si(e, t) {
  var n = ii(e) ? { first: e } : typeof e == "number" ? { each: e } : e, r = n.first, i = n.each, o = n.with, s = o === void 0 ? ai : o, a = n.scheduler, c = a === void 0 ? Wr : a, u = n.meta, l = u === void 0 ? null : u;
  if (r == null && i == null)
    throw new TypeError("No timeout provided.");
  return V(function(f, h) {
    var d, m, p = null, g = 0, v = function(S) {
      m = Z(h, c, function() {
        try {
          d.unsubscribe(), Y(s({
            meta: l,
            lastValue: p,
            seen: g
          })).subscribe(h);
        } catch (k) {
          h.error(k);
        }
      }, S);
    };
    d = f.subscribe(W(h, function(S) {
      m == null || m.unsubscribe(), g++, h.next(p = S), i > 0 && v(i);
    }, void 0, void 0, function() {
      m != null && m.closed || m == null || m.unsubscribe(), p = null;
    })), !g && v(r != null ? typeof r == "number" ? r : +r - c.now() : i);
  });
}
function ai(e) {
  throw new oi(e);
}
function A(e, t) {
  return V(function(n, r) {
    var i = 0;
    n.subscribe(W(r, function(o) {
      r.next(e.call(t, o, i++));
    }));
  });
}
var ci = Array.isArray;
function ui(e, t) {
  return ci(t) ? e.apply(void 0, Se([], he(t))) : e(t);
}
function li(e) {
  return A(function(t) {
    return ui(e, t);
  });
}
function fi(e, t, n, r, i, o, s, a) {
  var c = [], u = 0, l = 0, f = !1, h = function() {
    f && !c.length && !u && t.complete();
  }, d = function(p) {
    return u < r ? m(p) : c.push(p);
  }, m = function(p) {
    u++;
    var g = !1;
    Y(n(p, l++)).subscribe(W(t, function(v) {
      t.next(v);
    }, function() {
      g = !0;
    }, void 0, function() {
      if (g)
        try {
          u--;
          for (var v = function() {
            var S = c.shift();
            s || m(S);
          }; c.length && u < r; )
            v();
          h();
        } catch (S) {
          t.error(S);
        }
    }));
  };
  return e.subscribe(W(t, d, function() {
    f = !0, h();
  })), function() {
  };
}
function ht(e, t, n) {
  return n === void 0 && (n = 1 / 0), P(t) ? ht(function(r, i) {
    return A(function(o, s) {
      return t(r, o, i, s);
    })(Y(e(r, i)));
  }, n) : (typeof t == "number" && (n = t), V(function(r, i) {
    return fi(r, i, e, n);
  }));
}
function yn(e) {
  return e === void 0 && (e = 1 / 0), ht(ft, e);
}
function di() {
  return yn(1);
}
function De() {
  for (var e = [], t = 0; t < arguments.length; t++)
    e[t] = arguments[t];
  return di()(pt(e, Be(e)));
}
var pi = ["addListener", "removeListener"], hi = ["addEventListener", "removeEventListener"], vi = ["on", "off"];
function tt(e, t, n, r) {
  if (P(n) && (r = n, n = void 0), r)
    return tt(e, t, n).pipe(li(r));
  var i = he(bi(e) ? hi.map(function(a) {
    return function(c) {
      return e[a](t, c, n);
    };
  }) : mi(e) ? pi.map(Ot(e, t)) : gi(e) ? vi.map(Ot(e, t)) : [], 2), o = i[0], s = i[1];
  if (!o && dt(e))
    return ht(function(a) {
      return tt(a, t, n);
    })(Y(e));
  if (!o)
    throw new TypeError("Invalid event target");
  return new O(function(a) {
    var c = function() {
      for (var u = [], l = 0; l < arguments.length; l++)
        u[l] = arguments[l];
      return a.next(1 < u.length ? u : u[0]);
    };
    return o(c), function() {
      return s(c);
    };
  });
}
function Ot(e, t) {
  return function(n) {
    return function(r) {
      return e[n](t, r);
    };
  };
}
function mi(e) {
  return P(e.addListener) && P(e.removeListener);
}
function gi(e) {
  return P(e.on) && P(e.off);
}
function bi(e) {
  return P(e.addEventListener) && P(e.removeEventListener);
}
function yi() {
  for (var e = [], t = 0; t < arguments.length; t++)
    e[t] = arguments[t];
  var n = Be(e), r = Hr(e, 1 / 0), i = e;
  return i.length ? i.length === 1 ? Y(i[0]) : yn(r)(pt(i, n)) : Br;
}
function vt(e, t) {
  return V(function(n, r) {
    var i = 0;
    n.subscribe(W(r, function(o) {
      return e.call(t, o, i++) && r.next(o);
    }));
  });
}
function wi(e, t, n, r, i) {
  return function(o, s) {
    var a = n, c = t, u = 0;
    o.subscribe(W(s, function(l) {
      var f = u++;
      c = a ? e(c, l, f) : (a = !0, l), s.next(c);
    }, i));
  };
}
function Si(e, t) {
  return t === void 0 && (t = ft), e = e ?? xi, V(function(n, r) {
    var i, o = !0;
    n.subscribe(W(r, function(s) {
      var a = t(s);
      (o || !e(i, a)) && (o = !1, i = a, r.next(s));
    }));
  });
}
function xi(e, t) {
  return e === t;
}
function Ei(e) {
  return V(function(t, n) {
    try {
      t.subscribe(n);
    } finally {
      n.add(e);
    }
  });
}
function Ti(e, t) {
  return V(wi(e, t, arguments.length >= 2, !0));
}
function Ci(e) {
  return V(function(t, n) {
    var r = !1, i = W(n, function() {
      i == null || i.unsubscribe(), r = !0;
    }, Ie);
    Y(e).subscribe(i), t.subscribe(W(n, function(o) {
      return r && n.next(o);
    }));
  });
}
function M() {
  for (var e = [], t = 0; t < arguments.length; t++)
    e[t] = arguments[t];
  var n = Be(e);
  return V(function(r, i) {
    (n ? De(e, r, n) : De(e, r)).subscribe(i);
  });
}
function wn(e, t) {
  return V(function(n, r) {
    var i = null, o = 0, s = !1, a = function() {
      return s && !i && r.complete();
    };
    n.subscribe(W(r, function(c) {
      i == null || i.unsubscribe();
      var u = 0, l = o++;
      Y(e(c, l)).subscribe(i = W(r, function(f) {
        return r.next(t ? t(c, f, l, u++) : f);
      }, function() {
        i = null, a();
      }));
    }, function() {
      s = !0, a();
    }));
  });
}
function Lt(e) {
  return V(function(t, n) {
    Y(e).subscribe(W(n, function() {
      return n.complete();
    }, Ie)), !n.closed && t.subscribe(n);
  });
}
var Pi = Object.defineProperty, Ai = Object.defineProperties, ki = Object.getOwnPropertyDescriptors, _t = Object.getOwnPropertySymbols, Oi = Object.prototype.hasOwnProperty, Li = Object.prototype.propertyIsEnumerable, It = (e, t, n) => t in e ? Pi(e, t, { enumerable: !0, configurable: !0, writable: !0, value: n }) : e[t] = n, K = (e, t) => {
  for (var n in t || (t = {}))
    Oi.call(t, n) && It(e, n, t[n]);
  if (_t)
    for (var n of _t(t))
      Li.call(t, n) && It(e, n, t[n]);
  return e;
}, ye = (e, t) => Ai(e, ki(t)), B = (e, t, n) => new Promise((r, i) => {
  var o = (c) => {
    try {
      a(n.next(c));
    } catch (u) {
      i(u);
    }
  }, s = (c) => {
    try {
      a(n.throw(c));
    } catch (u) {
      i(u);
    }
  }, a = (c) => c.done ? r(c.value) : Promise.resolve(c.value).then(o, s);
  a((n = n.apply(e, t)).next());
}), Sn = "lk";
function j(e) {
  return typeof e > "u" ? !1 : _i(e) || Ii(e);
}
function _i(e) {
  var t;
  return e ? e.hasOwnProperty("participant") && e.hasOwnProperty("source") && e.hasOwnProperty("track") && typeof ((t = e.publication) == null ? void 0 : t.track) < "u" : !1;
}
function Ii(e) {
  return e ? e.hasOwnProperty("participant") && e.hasOwnProperty("source") && e.hasOwnProperty("publication") && typeof e.publication < "u" : !1;
}
function xe(e) {
  return e ? e.hasOwnProperty("participant") && e.hasOwnProperty("source") && typeof e.publication > "u" : !1;
}
function $(e) {
  if (typeof e == "string" || typeof e == "number")
    return `${e}`;
  if (xe(e))
    return `${e.participant.identity}_${e.source}_placeholder`;
  if (j(e))
    return `${e.participant.identity}_${e.publication.source}_${e.publication.trackSid}`;
  throw new Error(`Can't generate a id for the given track reference: ${e}`);
}
function Po(e, t) {
  return e === void 0 || t === void 0 ? !1 : j(e) && j(t) ? e.publication.trackSid === t.publication.trackSid : $(e) === $(t);
}
function Ao(e, t) {
  return typeof t > "u" ? !1 : j(e) ? t.some(
    (n) => n.participant.identity === e.participant.identity && j(n) && n.publication.trackSid === e.publication.trackSid
  ) : xe(e) ? t.some(
    (n) => n.participant.identity === e.participant.identity && xe(n) && n.source === e.source
  ) : !1;
}
function Di(e, t) {
  return xe(e) && j(t) && t.participant.identity === e.participant.identity && t.source === e.source;
}
function ko(e) {
  return e instanceof ot;
}
function Oo() {
  const e = document.createElement("p");
  e.style.width = "100%", e.style.height = "200px";
  const t = document.createElement("div");
  t.style.position = "absolute", t.style.top = "0px", t.style.left = "0px", t.style.visibility = "hidden", t.style.width = "200px", t.style.height = "150px", t.style.overflow = "hidden", t.appendChild(e), document.body.appendChild(t);
  const n = e.offsetWidth;
  t.style.overflow = "scroll";
  let r = e.offsetWidth;
  return n === r && (r = t.clientWidth), document.body.removeChild(t), n - r;
}
function Lo() {
  return typeof document < "u";
}
function Mi(e) {
  e = K({}, e);
  const t = "(?:(?:[a-z]+:)?//)?", n = "(?:\\S+(?::\\S*)?@)?", r = new RegExp(
    "(?:25[0-5]|2[0-4]\\d|1\\d\\d|[1-9]\\d|\\d)(?:\\.(?:25[0-5]|2[0-4]\\d|1\\d\\d|[1-9]\\d|\\d)){3}",
    "g"
  ).source, u = `(?:${t}|www\\.)${n}(?:localhost|${r}|(?:(?:[a-z\\u00a1-\\uffff0-9][-_]*)*[a-z\\u00a1-\\uffff0-9]+)(?:\\.(?:[a-z\\u00a1-\\uffff0-9]-*)*[a-z\\u00a1-\\uffff0-9]+)*(?:\\.(?:[a-z\\u00a1-\\uffff]{2,}))\\.?)(?::\\d{2,5})?(?:[/?#][^\\s"]*)?`;
  return e.exact ? new RegExp(`(?:^${u}$)`, "i") : new RegExp(u, "ig");
}
var Dt = "[^\\.\\s@:](?:[^\\s@:]*[^\\s@:\\.])?@[^\\.\\s@]+(?:\\.[^\\.\\s@]+)*";
function Ri({ exact: e } = {}) {
  return e ? new RegExp(`^${Dt}$`) : new RegExp(Dt, "g");
}
function _o(e, t) {
  return B(this, null, function* () {
    const { x: n, y: r } = yield yr(e, t, {
      placement: "top",
      middleware: [mr(6), br(), gr({ padding: 5 })]
    });
    return { x: n, y: r };
  });
}
function Io(e, t) {
  return !e.contains(t.target);
}
var Do = () => ({
  email: Ri(),
  url: Mi({})
});
function Mo(e, t) {
  const n = Object.entries(t).map(
    ([o, s], a) => Array.from(e.matchAll(s)).map(({ index: c, 0: u }) => ({
      type: o,
      weight: a,
      content: u,
      index: c ?? 0
    }))
  ).flat().sort((o, s) => {
    const a = o.index - s.index;
    return a !== 0 ? a : o.weight - s.weight;
  }).filter(({ index: o }, s, a) => {
    if (s === 0) return !0;
    const c = a[s - 1];
    return c.index + c.content.length <= o;
  }), r = [];
  let i = 0;
  for (const { type: o, content: s, index: a } of n)
    a > i && r.push(e.substring(i, a)), r.push({ type: o, content: s }), i = a + s.length;
  return e.length > i && r.push(e.substring(i)), r;
}
var Ni = [
  b.ConnectionStateChanged,
  b.RoomMetadataChanged,
  b.ActiveSpeakersChanged,
  b.ConnectionQualityChanged,
  b.ParticipantConnected,
  b.ParticipantDisconnected,
  b.ParticipantPermissionsChanged,
  b.ParticipantMetadataChanged,
  b.ParticipantNameChanged,
  b.ParticipantAttributesChanged,
  b.TrackMuted,
  b.TrackUnmuted,
  b.TrackPublished,
  b.TrackUnpublished,
  b.TrackStreamStateChanged,
  b.TrackSubscriptionFailed,
  b.TrackSubscriptionPermissionChanged,
  b.TrackSubscriptionStatusChanged
], xn = [
  ...Ni,
  b.LocalTrackPublished,
  b.LocalTrackUnpublished
], $i = [
  y.TrackPublished,
  y.TrackUnpublished,
  y.TrackMuted,
  y.TrackUnmuted,
  y.TrackStreamStateChanged,
  y.TrackSubscribed,
  y.TrackUnsubscribed,
  y.TrackSubscriptionPermissionChanged,
  y.TrackSubscriptionFailed,
  y.LocalTrackPublished,
  y.LocalTrackUnpublished
], Ui = [
  y.ConnectionQualityChanged,
  y.IsSpeakingChanged,
  y.ParticipantMetadataChanged,
  y.ParticipantPermissionsChanged,
  y.TrackMuted,
  y.TrackUnmuted,
  y.TrackPublished,
  y.TrackUnpublished,
  y.TrackStreamStateChanged,
  y.TrackSubscriptionFailed,
  y.TrackSubscriptionPermissionChanged,
  y.TrackSubscriptionStatusChanged
], En = [
  ...Ui,
  y.LocalTrackPublished,
  y.LocalTrackUnpublished
], L = Er.getLogger("lk-components-js");
L.setDefaultLevel("WARN");
function Ro(e, t = {}) {
  var n;
  L.setLevel(e), Fn((n = t.liveKitClientLogLevel) != null ? n : e);
}
function No(e, t = {}) {
  var n;
  const r = L.methodFactory;
  L.methodFactory = (i, o, s) => {
    const a = r(i, o, s), c = mt[i], u = c >= o && c < mt.silent;
    return (l, f) => {
      f ? a(l, f) : a(l), u && e(c, l, f);
    };
  }, L.setLevel(L.getLevel()), jn((n = t.liveKitClientLogExtension) != null ? n : e);
}
var $o = [
  {
    columns: 1,
    rows: 1
  },
  {
    columns: 1,
    rows: 2,
    orientation: "portrait"
  },
  {
    columns: 2,
    rows: 1,
    orientation: "landscape"
  },
  {
    columns: 2,
    rows: 2,
    minWidth: 560
  },
  {
    columns: 3,
    rows: 3,
    minWidth: 700
  },
  {
    columns: 4,
    rows: 4,
    minWidth: 960
  },
  {
    columns: 5,
    rows: 5,
    minWidth: 1100
  }
];
function Fi(e, t, n, r) {
  if (e.length < 1)
    throw new Error("At least one grid layout definition must be provided.");
  const i = ji(e);
  if (n <= 0 || r <= 0)
    return i[0];
  let o = 0;
  const s = n / r > 1 ? "landscape" : "portrait";
  let a = i.find((c, u, l) => {
    o = u;
    const f = l.findIndex((h, d) => {
      const m = !h.orientation || h.orientation === s, p = d > u, g = h.maxTiles === c.maxTiles;
      return p && g && m;
    }) !== -1;
    return c.maxTiles >= t && !f;
  });
  if (a === void 0)
    if (a = i[i.length - 1], a)
      L.warn(
        `No layout found for: participantCount: ${t}, width/height: ${n}/${r} fallback to biggest available layout (${a}).`
      );
    else
      throw new Error("No layout or fallback layout found.");
  if ((n < a.minWidth || r < a.minHeight) && o > 0) {
    const c = i[o - 1];
    a = Fi(
      i.slice(0, o),
      c.maxTiles,
      n,
      r
    );
  }
  return a;
}
function ji(e) {
  return [...e].map((t) => {
    var n, r;
    return {
      name: `${t.columns}x${t.rows}`,
      columns: t.columns,
      rows: t.rows,
      maxTiles: t.columns * t.rows,
      minWidth: (n = t.minWidth) != null ? n : 0,
      minHeight: (r = t.minHeight) != null ? r : 0,
      orientation: t.orientation
    };
  }).sort((t, n) => t.maxTiles !== n.maxTiles ? t.maxTiles - n.maxTiles : t.minWidth !== 0 || n.minWidth !== 0 ? t.minWidth - n.minWidth : t.minHeight !== 0 || n.minHeight !== 0 ? t.minHeight - n.minHeight : 0);
}
function Uo() {
  return typeof navigator < "u" && navigator.mediaDevices && !!navigator.mediaDevices.getDisplayMedia;
}
function Fo(e, t) {
  var n;
  return ye(K({}, e), {
    receivedAtMediaTimestamp: (n = t.rtpTimestamp) != null ? n : 0,
    receivedAt: t.timestamp
  });
}
function jo(e, t, n) {
  return [...e, ...t].reduceRight((r, i) => (r.find((o) => o.id === i.id) || r.unshift(i), r), []).slice(0 - n);
}
var Tn = [], Cn = {
  showChat: !1,
  unreadMessages: 0,
  showSettings: !1
};
function Wi(e) {
  return typeof e == "object";
}
function Wo(e) {
  return Array.isArray(e) && e.filter(Wi).length > 0;
}
function Pn(e, t) {
  return t.audioLevel - e.audioLevel;
}
function An(e, t) {
  return e.isSpeaking === t.isSpeaking ? 0 : e.isSpeaking ? -1 : 1;
}
function kn(e, t) {
  var n, r, i, o;
  return e.lastSpokeAt !== void 0 || t.lastSpokeAt !== void 0 ? ((r = (n = t.lastSpokeAt) == null ? void 0 : n.getTime()) != null ? r : 0) - ((o = (i = e.lastSpokeAt) == null ? void 0 : i.getTime()) != null ? o : 0) : 0;
}
function Me(e, t) {
  var n, r, i, o;
  return ((r = (n = e.joinedAt) == null ? void 0 : n.getTime()) != null ? r : 0) - ((o = (i = t.joinedAt) == null ? void 0 : i.getTime()) != null ? o : 0);
}
function Bi(e, t) {
  return j(e) ? j(t) ? 0 : -1 : j(t) ? 1 : 0;
}
function Vi(e, t) {
  const n = e.participant.isCameraEnabled, r = t.participant.isCameraEnabled;
  return n !== r ? n ? -1 : 1 : 0;
}
function Bo(e) {
  const t = [], n = [], r = [], i = [];
  e.forEach((a) => {
    a.participant.isLocal && a.source === R.Source.Camera ? t.push(a) : a.source === R.Source.ScreenShare ? n.push(a) : a.source === R.Source.Camera ? r.push(a) : i.push(a);
  });
  const o = Hi(n), s = zi(r);
  return [...t, ...o, ...s, ...i];
}
function Hi(e) {
  const t = [], n = [];
  return e.forEach((i) => {
    i.participant.isLocal ? t.push(i) : n.push(i);
  }), t.sort((i, o) => Me(i.participant, o.participant)), n.sort((i, o) => Me(i.participant, o.participant)), [...n, ...t];
}
function zi(e) {
  const t = [], n = [];
  return e.forEach((r) => {
    r.participant.isLocal ? t.push(r) : n.push(r);
  }), n.sort((r, i) => r.participant.isSpeaking && i.participant.isSpeaking ? Pn(r.participant, i.participant) : r.participant.isSpeaking !== i.participant.isSpeaking ? An(r.participant, i.participant) : r.participant.lastSpokeAt !== i.participant.lastSpokeAt ? kn(r.participant, i.participant) : j(r) !== j(i) ? Bi(r, i) : r.participant.isCameraEnabled !== i.participant.isCameraEnabled ? Vi(r, i) : Me(r.participant, i.participant)), [...t, ...n];
}
function Vo(e) {
  const t = [...e];
  t.sort((r, i) => {
    if (r.isSpeaking && i.isSpeaking)
      return Pn(r, i);
    if (r.isSpeaking !== i.isSpeaking)
      return An(r, i);
    if (r.lastSpokeAt !== i.lastSpokeAt)
      return kn(r, i);
    const o = r.videoTrackPublications.size > 0, s = i.videoTrackPublications.size > 0;
    return o !== s ? o ? -1 : 1 : Me(r, i);
  });
  const n = t.find((r) => r instanceof ot);
  if (n) {
    const r = t.indexOf(n);
    r >= 0 && (t.splice(r, 1), t.length > 0 ? t.splice(0, 0, n) : t.push(n));
  }
  return t;
}
function Yi(e, t) {
  return e.reduce(
    (n, r, i) => i % t === 0 ? [...n, [r]] : [...n.slice(0, -1), [...n.slice(-1)[0], r]],
    []
  );
}
function Mt(e, t) {
  const n = Math.max(e.length, t.length);
  return new Array(n).fill([]).map((r, i) => [e[i], t[i]]);
}
function Re(e, t, n) {
  return e.filter((r) => !t.map((i) => n(i)).includes(n(r)));
}
function nt(e) {
  return e.map((t) => typeof t == "string" || typeof t == "number" ? `${t}` : $(t));
}
function Gi(e, t) {
  return {
    dropped: Re(e, t, $),
    added: Re(t, e, $)
  };
}
function Ki(e) {
  return e.added.length !== 0 || e.dropped.length !== 0;
}
function rt(e, t) {
  const n = t.findIndex(
    (r) => $(r) === $(e)
  );
  if (n === -1)
    throw new Error(
      `Element not part of the array: ${$(
        e
      )} not in ${nt(t)}`
    );
  return n;
}
function Qi(e, t, n) {
  const r = rt(e, n), i = rt(t, n);
  return n.splice(r, 1, t), n.splice(i, 1, e), n;
}
function Ji(e, t) {
  const n = rt(e, t);
  return t.splice(n, 1), t;
}
function qi(e, t) {
  return [...t, e];
}
function Ge(e, t) {
  return Yi(e, t);
}
function Ho(e, t, n) {
  let r = Xi(e, t);
  if (r.length < t.length) {
    const s = Re(t, r, $);
    r = [...r, ...s];
  }
  const i = Ge(r, n), o = Ge(t, n);
  if (Mt(i, o).forEach(([s, a], c) => {
    if (s && a) {
      const u = Ge(r, n)[c], l = Gi(u, a);
      Ki(l) && (L.debug(
        `Detected visual changes on page: ${c}, current: ${nt(
          s
        )}, next: ${nt(a)}`,
        { changes: l }
      ), l.added.length === l.dropped.length && Mt(l.added, l.dropped).forEach(([f, h]) => {
        if (f && h)
          r = Qi(f, h, r);
        else
          throw new Error(
            `For a swap action we need a addition and a removal one is missing: ${f}, ${h}`
          );
      }), l.added.length === 0 && l.dropped.length > 0 && l.dropped.forEach((f) => {
        r = Ji(f, r);
      }), l.added.length > 0 && l.dropped.length === 0 && l.added.forEach((f) => {
        r = qi(f, r);
      }));
    }
  }), r.length > t.length) {
    const s = Re(r, t, $);
    r = r.filter(
      (a) => !s.map($).includes($(a))
    );
  }
  return r;
}
function Xi(e, t) {
  return e.map((n) => {
    const r = t.find(
      (i) => (
        // If the IDs match or ..
        $(n) === $(i) || // ... if the current item is a placeholder and the new item is the track reference can replace it.
        typeof n != "number" && xe(n) && j(i) && Di(n, i)
      )
    );
    return r ?? n;
  });
}
function U(e) {
  return `${Sn}-${e}`;
}
function zo(e) {
  const t = Rt(e), n = On(e.participant).pipe(
    A(() => Rt(e)),
    M(t)
  );
  return { className: U(
    e.source === R.Source.Camera || e.source === R.Source.ScreenShare ? "participant-media-video" : "participant-media-audio"
  ), trackObserver: n };
}
function Rt(e) {
  if (j(e))
    return e.publication;
  {
    const { source: t, name: n, participant: r } = e;
    if (t && n)
      return r.getTrackPublications().find((i) => i.source === t && i.trackName === n);
    if (n)
      return r.getTrackPublicationByName(n);
    if (t)
      return r.getTrackPublication(t);
    throw new Error("At least one of source and name needs to be defined");
  }
}
function se(e, ...t) {
  return new O((r) => {
    const i = () => {
      r.next(e);
    };
    return t.forEach((s) => {
      e.on(s, i);
    }), () => {
      t.forEach((s) => {
        e.off(s, i);
      });
    };
  }).pipe(M(e));
}
function ae(e, t) {
  return new O((r) => {
    const i = (...s) => {
      r.next(s);
    };
    return e.on(t, i), () => {
      e.off(t, i);
    };
  });
}
function Yo(e) {
  return ae(e, b.ConnectionStateChanged).pipe(
    A(([t]) => t),
    M(e.state)
  );
}
function Go(e) {
  return se(
    e,
    b.RoomMetadataChanged,
    b.ConnectionStateChanged
  ).pipe(
    A((n) => ({ name: n.name, metadata: n.metadata }))
  );
}
function Ko(e) {
  return ae(e, b.ActiveSpeakersChanged).pipe(
    A(([t]) => t)
  );
}
function Qo(e, t, n = !0) {
  var r;
  const i = () => B(this, null, function* () {
    try {
      const a = yield gt.getLocalDevices(e, n);
      o.next(a);
    } catch (a) {
      t == null || t(a);
    }
  }), o = new ne(), s = o.pipe(
    Ei(() => {
      var a;
      (a = navigator == null ? void 0 : navigator.mediaDevices) == null || a.removeEventListener("devicechange", i);
    })
  );
  if (typeof window < "u") {
    if (!window.isSecureContext)
      throw new Error(
        "Accessing media devices is available only in secure contexts (HTTPS and localhost), in some or all supporting browsers. See: https://developer.mozilla.org/en-US/docs/Web/API/Navigator/mediaDevices"
      );
    (r = navigator == null ? void 0 : navigator.mediaDevices) == null || r.addEventListener("devicechange", i);
  }
  return De(
    gt.getLocalDevices(e, n).catch((a) => (t == null || t(a), [])),
    s
  );
}
function Zi(e) {
  return ae(e, b.DataReceived);
}
function eo(e) {
  return ae(e, b.ChatMessage);
}
function to(e) {
  return se(e, b.AudioPlaybackStatusChanged).pipe(
    A((n) => ({ canPlayAudio: n.canPlaybackAudio }))
  );
}
function no(e) {
  return se(e, b.VideoPlaybackStatusChanged).pipe(
    A((n) => ({ canPlayVideo: n.canPlaybackVideo }))
  );
}
function ro(e, t) {
  return ae(e, b.ActiveDeviceChanged).pipe(
    vt(([n]) => n === t),
    A(([n, r]) => (L.debug("activeDeviceObservable | RoomEvent.ActiveDeviceChanged", { kind: n, deviceId: r }), r)),
    M(e.getActiveDevice(t))
  );
}
function Jo(e, t) {
  return ae(e, b.ParticipantEncryptionStatusChanged).pipe(
    vt(
      ([, n]) => (t == null ? void 0 : t.identity) === (n == null ? void 0 : n.identity) || !n && (t == null ? void 0 : t.identity) === e.localParticipant.identity
    ),
    A(([n]) => n),
    M(
      t instanceof ot ? t.isE2EEEnabled : !!(t != null && t.isEncrypted)
    )
  );
}
function qo(e) {
  return ae(e, b.RecordingStatusChanged).pipe(
    A(([t]) => t),
    M(e.isRecording)
  );
}
function me(e, ...t) {
  return new O((r) => {
    const i = () => {
      r.next(e);
    };
    return t.forEach((s) => {
      e.on(s, i);
    }), () => {
      t.forEach((s) => {
        e.off(s, i);
      });
    };
  }).pipe(M(e));
}
function On(e) {
  return me(
    e,
    y.TrackMuted,
    y.TrackUnmuted,
    y.ParticipantPermissionsChanged,
    // ParticipantEvent.IsSpeakingChanged,
    y.TrackPublished,
    y.TrackUnpublished,
    y.LocalTrackPublished,
    y.LocalTrackUnpublished,
    y.MediaDevicesError,
    y.TrackSubscriptionStatusChanged
    // ParticipantEvent.ConnectionQualityChanged,
  ).pipe(
    A((n) => {
      const { isMicrophoneEnabled: r, isCameraEnabled: i, isScreenShareEnabled: o } = n, s = n.getTrackPublication(R.Source.Microphone), a = n.getTrackPublication(R.Source.Camera);
      return {
        isCameraEnabled: i,
        isMicrophoneEnabled: r,
        isScreenShareEnabled: o,
        cameraTrack: a,
        microphoneTrack: s,
        participant: n
      };
    })
  );
}
function io(e) {
  return e ? me(
    e,
    y.ParticipantMetadataChanged,
    y.ParticipantNameChanged
  ).pipe(
    A(({ name: n, identity: r, metadata: i }) => ({
      name: n,
      identity: r,
      metadata: i
    })),
    M({
      name: e.name,
      identity: e.identity,
      metadata: e.metadata
    })
  ) : void 0;
}
function oo(e) {
  return Ve(
    e,
    y.ConnectionQualityChanged
  ).pipe(
    A(([n]) => n),
    M(e.connectionQuality)
  );
}
function Ve(e, t) {
  return new O((r) => {
    const i = (...s) => {
      r.next(s);
    };
    return e.on(t, i), () => {
      e.off(t, i);
    };
  });
}
function so(e) {
  var t, n, r, i;
  return me(
    e.participant,
    y.TrackMuted,
    y.TrackUnmuted,
    y.TrackSubscribed,
    y.TrackUnsubscribed,
    y.LocalTrackPublished,
    y.LocalTrackUnpublished
  ).pipe(
    A((o) => {
      var s, a;
      const c = (s = e.publication) != null ? s : o.getTrackPublication(e.source);
      return (a = c == null ? void 0 : c.isMuted) != null ? a : !0;
    }),
    M(
      (i = (r = (t = e.publication) == null ? void 0 : t.isMuted) != null ? r : (n = e.participant.getTrackPublication(e.source)) == null ? void 0 : n.isMuted) != null ? i : !0
    )
  );
}
function Xo(e) {
  return Ve(e, y.IsSpeakingChanged).pipe(
    A(([t]) => t)
  );
}
function Zo(e, t = {}) {
  var n;
  let r;
  const i = new O((c) => (r = c, () => a.unsubscribe())).pipe(M(Array.from(e.remoteParticipants.values()))), o = (n = t.additionalRoomEvents) != null ? n : xn, s = Array.from(
    /* @__PURE__ */ new Set([
      b.ParticipantConnected,
      b.ParticipantDisconnected,
      b.ConnectionStateChanged,
      ...o
    ])
  ), a = se(e, ...s).subscribe(
    (c) => r == null ? void 0 : r.next(Array.from(c.remoteParticipants.values()))
  );
  return e.remoteParticipants.size > 0 && (r == null || r.next(Array.from(e.remoteParticipants.values()))), i;
}
function es(e, t, n = {}) {
  var r;
  const i = (r = n.additionalEvents) != null ? r : En;
  return se(
    e,
    b.ParticipantConnected,
    b.ParticipantDisconnected,
    b.ConnectionStateChanged
  ).pipe(
    wn((s) => {
      const a = s.getParticipantByIdentity(t);
      return a ? me(a, ...i) : new O((c) => c.next(void 0));
    }),
    M(e.getParticipantByIdentity(t))
  );
}
function ts(e) {
  return Ve(
    e,
    y.ParticipantPermissionsChanged
  ).pipe(
    A(() => e.permissions),
    M(e.permissions)
  );
}
function ns(e, { kind: t, identity: n }, r = {}) {
  var i;
  const o = (i = r.additionalEvents) != null ? i : En, s = (c) => {
    let u = !0;
    return t && (u = u && c.kind === t), n && (u = u && c.identity === n), u;
  };
  return se(
    e,
    b.ParticipantConnected,
    b.ParticipantDisconnected,
    b.ConnectionStateChanged
  ).pipe(
    wn((c) => {
      const u = Array.from(c.remoteParticipants.values()).find(
        (l) => s(l)
      );
      return u ? me(u, ...o) : new O((l) => l.next(void 0));
    }),
    M(Array.from(e.remoteParticipants.values()).find((c) => s(c)))
  );
}
function rs(e) {
  return typeof e > "u" ? new O() : Ve(e, y.AttributesChanged).pipe(
    A(([t]) => ({
      changed: t,
      attributes: e.attributes
    })),
    M({ changed: e.attributes, attributes: e.attributes })
  );
}
function is(e, t, n, r, i) {
  const { localParticipant: o } = t, s = (f, h) => {
    let d = !1;
    switch (f) {
      case R.Source.Camera:
        d = h.isCameraEnabled;
        break;
      case R.Source.Microphone:
        d = h.isMicrophoneEnabled;
        break;
      case R.Source.ScreenShare:
        d = h.isScreenShareEnabled;
        break;
    }
    return d;
  }, a = On(o).pipe(
    A((f) => s(e, f.participant)),
    M(s(e, o))
  ), c = new ne(), u = (f, h) => B(this, null, function* () {
    try {
      switch (h ?? (h = n), c.next(!0), e) {
        case R.Source.Camera:
          return yield o.setCameraEnabled(
            f ?? !o.isCameraEnabled,
            h,
            r
          ), o.isCameraEnabled;
        case R.Source.Microphone:
          return yield o.setMicrophoneEnabled(
            f ?? !o.isMicrophoneEnabled,
            h,
            r
          ), o.isMicrophoneEnabled;
        case R.Source.ScreenShare:
          return yield o.setScreenShareEnabled(
            f ?? !o.isScreenShareEnabled,
            h,
            r
          ), o.isScreenShareEnabled;
        default:
          throw new TypeError("Tried to toggle unsupported source");
      }
    } catch (d) {
      if (i && d instanceof Error) {
        i == null || i(d);
        return;
      } else
        throw d;
    } finally {
      c.next(!1);
    }
  });
  return {
    className: U("button"),
    toggle: u,
    enabledObserver: a,
    pendingObserver: c.asObservable()
  };
}
function os() {
  let e = !1;
  const t = new ne(), n = new ne(), r = (o) => B(this, null, function* () {
    n.next(!0), e = o ?? !e, t.next(e), n.next(!1);
  });
  return {
    className: U("button"),
    toggle: r,
    enabledObserver: t.asObservable(),
    pendingObserver: n.asObservable()
  };
}
function ss(e, t, n) {
  const r = new on(void 0), i = t ? ro(t, e) : r.asObservable(), o = (a, ...c) => B(this, [a, ...c], function* (u, l = {}) {
    var f, h, d;
    if (t) {
      L.debug(`Switching active device of kind "${e}" with id ${u}.`), yield t.switchActiveDevice(e, u, l.exact);
      const m = (f = t.getActiveDevice(e)) != null ? f : u;
      m !== u && u !== "default" && L.info(
        `We tried to select the device with id (${u}), but the browser decided to select the device with id (${m}) instead.`
      );
      let p;
      e === "audioinput" ? p = (h = t.localParticipant.getTrackPublication(R.Source.Microphone)) == null ? void 0 : h.track : e === "videoinput" && (p = (d = t.localParticipant.getTrackPublication(R.Source.Camera)) == null ? void 0 : d.track);
      const g = u === "default" && !p || u === "default" && (p == null ? void 0 : p.mediaStreamTrack.label.startsWith("Default"));
      r.next(g ? u : m);
    } else if (n) {
      yield n.setDeviceId(l.exact ? { exact: u } : u);
      const m = yield n.getDeviceId();
      r.next(
        u === "default" && n.mediaStreamTrack.label.startsWith("Default") ? u : m
      );
    } else r.value !== u && (L.warn(
      "device switch skipped, please provide either a room or a local track to switch on. "
    ), r.next(u));
  });
  return {
    className: U("media-device-select"),
    activeDeviceObservable: i,
    setActiveMediaDevice: o
  };
}
function as(e) {
  const t = (r) => {
    e.disconnect(r);
  };
  return { className: U("disconnect-button"), disconnect: t };
}
function cs(e) {
  const t = U("connection-quality"), n = oo(e);
  return { className: t, connectionQualityObserver: n };
}
function us(e) {
  let t = "track-muted-indicator-camera";
  switch (e.source) {
    case R.Source.Camera:
      t = "track-muted-indicator-camera";
      break;
    case R.Source.Microphone:
      t = "track-muted-indicator-microphone";
      break;
  }
  const n = U(t), r = so(e);
  return { className: n, mediaMutedObserver: r };
}
function ls(e) {
  return { className: "lk-participant-name", infoObserver: io(e) };
}
function fs() {
  return {
    className: U("participant-tile")
  };
}
var Nt = {
  CHAT: "lk-chat-topic",
  CHAT_UPDATE: "lk-chat-update-topic"
};
function it(e, t) {
  return B(this, arguments, function* (n, r, i = {}) {
    const { reliable: o, destinationIdentities: s, topic: a } = i;
    yield n.publishData(r, {
      destinationIdentities: s,
      topic: a,
      reliable: o
    });
  });
}
function ao(e, t, n) {
  const r = Array.isArray(t) ? t : [t], i = Zi(e).pipe(
    vt(
      ([, , , c]) => t === void 0 || c !== void 0 && r.includes(c)
    ),
    A(([c, u, , l]) => {
      const f = {
        payload: c,
        topic: l,
        from: u
      };
      return n == null || n(f), f;
    })
  );
  let o;
  const s = new O((c) => {
    o = c;
  });
  return { messageObservable: i, isSendingObservable: s, send: (c, ...u) => B(this, [c, ...u], function* (l, f = {}) {
    o.next(!0);
    try {
      yield it(e.localParticipant, l, K({ topic: r[0] }, f));
    } finally {
      o.next(!1);
    }
  }) };
}
function co(e) {
  return { chatObservable: eo(e), send: (i) => B(this, null, function* () {
    return yield e.localParticipant.sendChatMessage(i);
  }), edit: (i, o) => B(this, null, function* () {
    return yield e.localParticipant.editChatMessage(i, o);
  }) };
}
var uo = new TextEncoder(), lo = new TextDecoder(), Pe = /* @__PURE__ */ new Map(), fo = (e) => uo.encode(JSON.stringify(e)), po = (e) => JSON.parse(lo.decode(e));
function ds(e, t) {
  var n, r;
  const i = new ne(), { messageDecoder: o, messageEncoder: s, channelTopic: a, updateChannelTopic: c } = t ?? {}, u = a ?? Nt.CHAT, l = c ?? Nt.CHAT_UPDATE;
  let f = !1;
  Pe.has(e) || (f = !0);
  const h = (n = Pe.get(e)) != null ? n : /* @__PURE__ */ new Map(), d = (r = h.get(u)) != null ? r : new ne();
  if (h.set(u, d), Pe.set(e, h), f) {
    const { messageObservable: C } = ao(e, [u, l]);
    C.pipe(Lt(i)).subscribe(d);
  }
  const { chatObservable: m, send: p } = co(e), g = o ?? po, v = yi(
    d.pipe(
      A((C) => {
        const T = g(C.payload), I = ye(K({}, T), { from: C.from });
        if (!ho(I))
          return I;
      })
    ),
    m.pipe(
      A(([C, T]) => ye(K({}, C), { from: T }))
    )
  ).pipe(
    Ti((C, T) => {
      var I;
      if (!T)
        return C;
      if ("id" in T && C.find((N) => {
        var w, x;
        return ((w = N.from) == null ? void 0 : w.identity) === ((x = T.from) == null ? void 0 : x.identity) && N.id === T.id;
      })) {
        const N = C.findIndex((w) => w.id === T.id);
        if (N > -1) {
          const w = C[N];
          C[N] = ye(K({}, T), {
            timestamp: w.timestamp,
            editTimestamp: (I = T.editTimestamp) != null ? I : T.timestamp
          });
        }
        return [...C];
      }
      return [...C, T];
    }, []),
    Lt(i)
  ), S = new on(!1), k = s ?? fo, _ = (C) => B(this, null, function* () {
    S.next(!0);
    try {
      const T = yield p(C), I = k(ye(K({}, T), { ignore: !0 }));
      return yield it(e.localParticipant, I, {
        reliable: !0,
        topic: u
      }), T;
    } finally {
      S.next(!1);
    }
  }), E = (C, T) => B(this, null, function* () {
    const I = Date.now(), N = typeof T == "string" ? { id: T, message: "", timestamp: I } : T;
    S.next(!0);
    try {
      const w = yield e.localParticipant.editChatMessage(C, N), x = k(w);
      return yield it(e.localParticipant, x, {
        topic: l,
        reliable: !0
      }), w;
    } finally {
      S.next(!1);
    }
  });
  function re() {
    i.next(), i.complete(), Pe.delete(e);
  }
  return e.once(b.Disconnected, re), {
    messageObservable: v,
    isSendingObservable: S,
    send: _,
    update: E
  };
}
function ho(e) {
  return e.ignore == !0;
}
function ps() {
  const e = (n) => B(this, null, function* () {
    L.info("Start Audio for room: ", n), yield n.startAudio();
  });
  return { className: U("start-audio-button"), roomAudioPlaybackAllowedObservable: to, handleStartAudioPlayback: e };
}
function hs() {
  const e = (n) => B(this, null, function* () {
    L.info("Start Video for room: ", n), yield n.startVideo();
  });
  return { className: U("start-audio-button"), roomVideoPlaybackAllowedObservable: no, handleStartVideoPlayback: e };
}
function vs() {
  return { className: [U("button"), U("chat-toggle")].join(" ") };
}
function ms() {
  return { className: [U("button"), U("focus-toggle-button")].join(" ") };
}
function gs() {
  return { className: "lk-clear-pin-button lk-button" };
}
function bs() {
  return { className: "lk-room-container" };
}
function $t(e, t, n = !0) {
  const i = [e.localParticipant, ...Array.from(e.remoteParticipants.values())], o = [];
  return i.forEach((s) => {
    t.forEach((a) => {
      const c = Array.from(
        s.trackPublications.values()
      ).filter(
        (u) => u.source === a && // either return all or only the ones that are subscribed
        (!n || u.track)
      ).map((u) => ({
        participant: s,
        publication: u,
        source: u.source
      }));
      o.push(...c);
    });
  }), { trackReferences: o, participants: i };
}
function Ut(e, t, n = !1) {
  const { sources: r, kind: i, name: o } = t;
  return Array.from(e.trackPublications.values()).filter(
    (a) => (!r || r.includes(a.source)) && (!i || a.kind === i) && (!o || a.trackName === o) && // either return all or only the ones that are subscribed
    (!n || a.track)
  ).map((a) => ({
    participant: e,
    publication: a,
    source: a.source
  }));
}
function ys(e, t, n) {
  var r, i;
  const o = (r = n.additionalRoomEvents) != null ? r : xn, s = (i = n.onlySubscribed) != null ? i : !0, a = Array.from(
    (/* @__PURE__ */ new Set([
      b.ParticipantConnected,
      b.ParticipantDisconnected,
      b.ConnectionStateChanged,
      b.LocalTrackPublished,
      b.LocalTrackUnpublished,
      b.TrackPublished,
      b.TrackUnpublished,
      b.TrackSubscriptionStatusChanged,
      ...o
    ])).values()
  );
  return se(e, ...a).pipe(
    A((u) => {
      const l = $t(u, t, s);
      return L.debug(`TrackReference[] was updated. (length ${l.trackReferences.length})`, l), l;
    }),
    M($t(e, t, s))
  );
}
function ws(e, t) {
  return me(e, ...$i).pipe(
    A((r) => {
      const i = Ut(r, t);
      return L.debug(`TrackReference[] was updated. (length ${i.length})`, i), i;
    }),
    M(Ut(e, t))
  );
}
function Ln(e, t) {
  return new O((r) => {
    const i = (...s) => {
      r.next(s);
    };
    return e.on(t, i), () => {
      e.off(t, i);
    };
  });
}
function Ss(e) {
  return Ln(e, Ft.TranscriptionReceived);
}
function xs(e) {
  return Ln(e, Ft.TimeSyncUpdate).pipe(
    A(([t]) => t)
  );
}
function Es(e, t = 1e3) {
  if (e === null) return kt(!1);
  const n = tt(e, "mousemove", { passive: !0 }).pipe(A(() => !0)), r = n.pipe(
    si({
      each: t,
      with: () => De(kt(!1), r.pipe(Ci(n)))
    }),
    Si()
  );
  return r;
}
function vo(e, t) {
  if (typeof localStorage > "u") {
    L.error("Local storage is not available.");
    return;
  }
  try {
    if (t) {
      const n = Object.fromEntries(
        Object.entries(t).filter(([, r]) => r !== "")
      );
      localStorage.setItem(e, JSON.stringify(n));
    }
  } catch (n) {
    L.error(`Error setting item to local storage: ${n}`);
  }
}
function mo(e) {
  if (typeof localStorage > "u") {
    L.error("Local storage is not available.");
    return;
  }
  try {
    const t = localStorage.getItem(e);
    if (!t) {
      L.warn(`Item with key ${e} does not exist in local storage.`);
      return;
    }
    return JSON.parse(t);
  } catch (t) {
    L.error(`Error getting item from local storage: ${t}`);
    return;
  }
}
function go(e) {
  return {
    load: () => mo(e),
    save: (t) => vo(e, t)
  };
}
var bo = `${Sn}-user-choices`, be = {
  videoEnabled: !0,
  audioEnabled: !0,
  videoDeviceId: "",
  audioDeviceId: "",
  username: ""
}, { load: yo, save: wo } = go(bo);
function Ts(e, t = !1) {
  t !== !0 && wo(e);
}
function Cs(e, t = !1) {
  var n, r, i, o, s;
  const a = {
    videoEnabled: (n = e == null ? void 0 : e.videoEnabled) != null ? n : be.videoEnabled,
    audioEnabled: (r = e == null ? void 0 : e.audioEnabled) != null ? r : be.audioEnabled,
    videoDeviceId: (i = e == null ? void 0 : e.videoDeviceId) != null ? i : be.videoDeviceId,
    audioDeviceId: (o = e == null ? void 0 : e.audioDeviceId) != null ? o : be.audioDeviceId,
    username: (s = e == null ? void 0 : e.username) != null ? s : be.username
  };
  if (t)
    return a;
  {
    const c = yo();
    return K(K({}, a), c ?? {});
  }
}
function _n(e, t) {
  if (t.msg === "show_chat")
    return { ...e, showChat: !0, unreadMessages: 0 };
  if (t.msg === "hide_chat")
    return { ...e, showChat: !1 };
  if (t.msg === "toggle_chat") {
    const n = { ...e, showChat: !e.showChat };
    return n.showChat === !0 && (n.unreadMessages = 0), n;
  } else return t.msg === "unread_msg" ? { ...e, unreadMessages: t.count } : t.msg === "toggle_settings" ? { ...e, showSettings: !e.showSettings } : { ...e };
}
function In(e, t) {
  return t.msg === "set_pin" ? [t.trackReference] : t.msg === "clear_pin" ? [] : { ...e };
}
const Dn = D.createContext(void 0);
function Ps() {
  const e = D.useContext(Dn);
  if (!e)
    throw Error("Tried to access LayoutContext context outside a LayoutContextProvider provider.");
  return e;
}
function As(e) {
  const t = So();
  if (e ?? (e = t), !e)
    throw Error("Tried to access LayoutContext context outside a LayoutContextProvider provider.");
  return e;
}
function ks() {
  const [e, t] = D.useReducer(In, Tn), [n, r] = D.useReducer(_n, Cn);
  return {
    pin: { dispatch: t, state: e },
    widget: { dispatch: r, state: n }
  };
}
function Os(e) {
  const [t, n] = D.useReducer(In, Tn), [r, i] = D.useReducer(_n, Cn);
  return e ?? {
    pin: { dispatch: n, state: t },
    widget: { dispatch: i, state: r }
  };
}
function So() {
  return D.useContext(Dn);
}
const Mn = D.createContext(
  void 0
);
function Ls() {
  const e = D.useContext(Mn);
  if (!e)
    throw Error("tried to access track context outside of track context provider");
  return e;
}
function Rn() {
  return D.useContext(Mn);
}
function _s(e) {
  const t = Rn(), n = e ?? t;
  if (!n)
    throw new Error(
      "No TrackRef, make sure you are inside a TrackRefContext or pass the TrackRef explicitly"
    );
  return n;
}
const Nn = D.createContext(void 0);
function Is() {
  const e = D.useContext(Nn);
  if (!e)
    throw Error("tried to access participant context outside of participant context provider");
  return e;
}
function xo() {
  return D.useContext(Nn);
}
function Ds(e) {
  const t = xo(), n = Rn(), r = e ?? t ?? (n == null ? void 0 : n.participant);
  if (!r)
    throw new Error(
      "No participant provided, make sure you are inside a participant context or pass the participant explicitly"
    );
  return r;
}
const $n = D.createContext(void 0);
function Ms() {
  const e = D.useContext($n);
  if (!e)
    throw Error("tried to access room context outside of livekit room component");
  return e;
}
function Eo() {
  return D.useContext($n);
}
function Rs(e) {
  const t = Eo(), n = e ?? t;
  if (!n)
    throw new Error(
      "No room provided, make sure you are inside a Room context or pass the room explicitly"
    );
  return n;
}
const To = D.createContext(void 0);
function Ns(e) {
  const t = D.useContext(To);
  if (e === !0) {
    if (t)
      return t;
    throw Error("tried to access feature context, but none is present");
  }
  return t;
}
export {
  Ts as $,
  Ho as A,
  xo as B,
  io as C,
  fs as D,
  Zo as E,
  As as F,
  $o as G,
  es as H,
  ns as I,
  Go as J,
  Ko as K,
  To as L,
  Vo as M,
  ps as N,
  hs as O,
  vs as P,
  us as Q,
  $n as R,
  is as S,
  os as T,
  Wi as U,
  ys as V,
  Wo as W,
  Rt as X,
  zo as Y,
  ds as Z,
  Cs as _,
  Ps as a,
  Jo as a0,
  j as a1,
  ws as a2,
  xs as a3,
  Ss as a4,
  jo as a5,
  Fo as a6,
  rs as a7,
  qo as a8,
  Rn as a9,
  Dn as aa,
  ls as ab,
  ko as ac,
  wr as ad,
  Sr as ae,
  Nn as af,
  Mn as ag,
  Ns as ah,
  Es as ai,
  Oo as aj,
  Os as ak,
  Mo as al,
  Do as am,
  _o as an,
  Io as ao,
  be as ap,
  Uo as aq,
  ks as ar,
  Po as as,
  Lo as at,
  Ro as au,
  No as av,
  Is as aw,
  Ls as ax,
  gs as b,
  Ds as c,
  cs as d,
  Yo as e,
  Ms as f,
  ao as g,
  as as h,
  _s as i,
  So as j,
  ms as k,
  L as l,
  Ao as m,
  Fi as n,
  so as o,
  $ as p,
  Xo as q,
  to as r,
  bs as s,
  On as t,
  Rs as u,
  ts as v,
  Eo as w,
  Qo as x,
  ss as y,
  Bo as z
};
